From 7bd9fd704dd3cea530174c127bf9a93e55d93b5e Mon Sep 17 00:00:00 2001
From: chandavath <rcrajagopal@habana.ai>
Date: Thu, 30 May 2024 19:36:49 +0300
Subject: [PATCH] [SW-165981] emb cmn: deleted common function and files

Deleted the common functions and files between legacy preboot
and preboot rtos and moved it to emb common mods folder.

Included emb common mods files for baremetal compilation.

Change-Id: I1c62aec467c35862752ee4a044357e2b73626640
Signed-off-by: chandavath <rcrajagopal@habana.ai>
---

diff --git a/Makefile b/Makefile
index 7e736e0..809bfce 100644
--- a/Makefile
+++ b/Makefile
@@ -10,12 +10,18 @@
 HL_DEV_ARCH := arch/$(BASE_DEV_NAME)/$(HL_DEV_SUBTYPE)
 HL_DEV_ARCH_COMMON := arch/$(BASE_DEV_NAME)/common
 
+EMB_COMMON_MOD_DIR := ${PREBOOT_ROOT_DIR}/emb_common_mod
+
 CFLAGS += -I$(SPECS_DIR)
 CFLAGS += -I$(SPECS_DIR)/$(BASE_DEV_NAME)/asic_reg
 CFLAGS += -I ${EMBEDDED_ROOT}/arcboot/arch
 CFLAGS += -Wall -Werror
 CFLAGS += -g -c -fno-builtin -fno-stack-protector
 CFLAGS += -I ./include/ -I ./ -I ./lib/ -I ./lib/common/ -I ./$(HL_DEV_ARCH)/ -I ./$(HL_DEV_ARCH_COMMON)/
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/include
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib/common
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib/preboot
 
 ifeq ($(HL_TARGET),pldm)
 	CFLAGS += -DTARGET_IS_PLDM
@@ -41,6 +47,8 @@
 PPBOOT_SRCS_C += $(shell ls app/ppboot/*.c)
 PPBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH)/common.c)
 PPBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH)/$(HL_TARGET).c)
+PPBOOT_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/lib/common/lib_common.c)
+PPBOOT_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/lib/common/pci.c)
 PPBOOT_APP_S := $(shell ls app/ppboot/crt0_ppboot.S)
 PPBOOT_APP_S += $(shell ls $(HL_DEV_ARCH)/*.S | grep -v lds)
 
diff --git a/app/ext_rom/Makefile b/app/ext_rom/Makefile
index d00bce2..bacabd5 100644
--- a/app/ext_rom/Makefile
+++ b/app/ext_rom/Makefile
@@ -7,6 +7,8 @@
 HL_DEV_ARCH_COMMON := $(PREBOOT_ROOT_DIR)/arch/$(BASE_DEV_NAME)/common
 SPECS_DEV := $(shell echo $(HL_DEV) | cut -d'_' -f1)
 
+EMB_COMMON_MOD_DIR := ${PREBOOT_ROOT_DIR}/emb_common_mod
+
 CFLAGS += -I$(SPECS_DIR)
 CFLAGS += -I$(SPECS_EMBEDDED_DIR)
 CFLAGS += -I$(SPECS_DIR)/$(BASE_DEV_NAME)/asic_reg
@@ -17,6 +19,10 @@
 CFLAGS += -include $(SPECS_EMBEDDED_DIR)/$(SPECS_DEV)/$(SPECS_DEV)_reg_map.h
 CFLAGS += -include erom_app_defs.h
 CFLAGS += -I ${EMBEDDED_ROOT}/arcboot/arch
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/include
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib/common
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib/preboot
 
 CFLAGS += -Os -fdata-sections -ffunction-sections
 SFLAGS := $(CFLAGS) -D__ASSEMBLY__
@@ -38,16 +44,16 @@
 
 EROM_PCIE_DEPS := \
 	$(PREBOOT_ROOT_DIR)/lib/common/pll.o \
-	$(PREBOOT_ROOT_DIR)/lib/common/pci.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/uart.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/boot_if.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/spi.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/spi_flash.o \
-	$(PREBOOT_ROOT_DIR)/lib/common/lib_common.o \
 	$(HL_DEV_ARCH)/$(HL_TARGET).o \
 	$(HL_DEV_ARCH)/../common/$(HL_TARGET).o \
 	$(HL_DEV_ARCH)/../common/common.o \
 	$(HL_DEV_ARCH)/common.o \
+	$(EMB_COMMON_MOD_DIR)/lib/common/lib_common.o \
+	$(EMB_COMMON_MOD_DIR)/lib/common/pci.o \
 	erom_secondary.o
 
 EROM_RECOVERY_DEPS := \
@@ -63,21 +69,16 @@
 	$(PREBOOT_ROOT_DIR)/lib/common/sha256.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/sha512.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/string.o \
-	$(PREBOOT_ROOT_DIR)/lib/common/pci.o \
-	$(PREBOOT_ROOT_DIR)/lib/common/lib_common.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/module_pci.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/boot_info.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/crc32.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/security.o \
-	$(PREBOOT_ROOT_DIR)/lib/preboot/fit_regions.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/module_fwupd.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/module_uart.o \
-	$(PREBOOT_ROOT_DIR)/lib/preboot/fit_rsa.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/module_i2c_slave.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/crc8.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/i2cslave_indirect.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/thermal.o \
-	$(PREBOOT_ROOT_DIR)/lib/preboot/fit.o \
 	$(PREBOOT_ROOT_DIR)/lib/preboot/memory.o \
 	$(HL_DEV_ARCH)/common.o \
 	$(HL_DEV_ARCH)/../common/common.o \
@@ -86,7 +87,13 @@
 	$(HL_DEV_ARCH)/../common/sha.o \
 	$(HL_DEV_ARCH)/../common/$(HL_TARGET).o \
 	$(HL_DEV_ARCH_COMMON)/reset.o \
-	$(HL_DEV_ARCH)/$(HL_TARGET).o
+	$(HL_DEV_ARCH)/$(HL_TARGET).o \
+	$(EMB_COMMON_MOD_DIR)/lib/common/lib_common.o \
+	$(EMB_COMMON_MOD_DIR)/lib/common/pci.o \
+	$(EMB_COMMON_MOD_DIR)/lib/preboot/fit_regions.o \
+	$(EMB_COMMON_MOD_DIR)/lib/preboot/fit_rsa.o \
+	$(EMB_COMMON_MOD_DIR)/lib/preboot/fit.o \
+	$(EMB_COMMON_MOD_DIR)/lib/preboot/fit_load.o
 
 include $(HL_DEV_ARCH)/Makefile
 
diff --git a/app/pcie_pif/Makefile b/app/pcie_pif/Makefile
index a3f4650..450a792 100644
--- a/app/pcie_pif/Makefile
+++ b/app/pcie_pif/Makefile
@@ -6,6 +6,8 @@
 HL_DEV_ARCH := $(PREBOOT_ROOT_DIR)/arch/$(BASE_DEV_NAME)/$(HL_DEV_SUBTYPE)
 HL_DEV_ARCH_COMMON := $(PREBOOT_ROOT_DIR)/arch/$(BASE_DEV_NAME)/common
 
+EMB_COMMON_MOD_DIR := ${PREBOOT_ROOT_DIR}/emb_common_mod
+
 CFLAGS += -I$(SPECS_DIR)
 CFLAGS += -I$(SPECS_EMBEDDED_DIR)
 CFLAGS += -I$(SPECS_EMBEDDED_DIR)/$(BASE_DEV_NAME)
@@ -13,6 +15,10 @@
 CFLAGS += -Wall -Werror
 CFLAGS += -g -c
 CFLAGS += -I $(PREBOOT_ROOT_DIR)/include/ -I $(PREBOOT_ROOT_DIR)/ -I $(PREBOOT_ROOT_DIR)/lib/ -I $(PREBOOT_ROOT_DIR)/lib/common/ -I $(HL_DEV_ARCH)/
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/include
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib/common
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib/preboot
 
 CFLAGS += -Os -fdata-sections -ffunction-sections
 
@@ -28,6 +34,8 @@
 PIF_ARC_SRCS_C += $(shell ls $(HL_DEV_ARCH_COMMON)/common.c)
 PIF_ARC_SRCS_C += $(shell ls $(HL_DEV_ARCH)/common.c)
 PIF_ARC_SRCS_C += $(shell ls $(HL_DEV_ARCH)/$(HL_TARGET).c)
+PIF_ARC_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/lib/common/lib_common.c)
+PIF_ARC_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/lib/common/pci.c)
 PIF_ARC_APP_S := $(shell ls crt0_pcie_pif.S)
 
 include $(HL_DEV_ARCH)/Makefile
diff --git a/app/pldm_bloader/Makefile b/app/pldm_bloader/Makefile
index ff94d37..c8ce6e3 100644
--- a/app/pldm_bloader/Makefile
+++ b/app/pldm_bloader/Makefile
@@ -6,6 +6,8 @@
 HL_DEV_ARCH := $(PREBOOT_ROOT_DIR)/arch/$(BASE_DEV_NAME)/$(HL_DEV_SUBTYPE)
 HL_DEV_ARCH_COMMON := $(PREBOOT_ROOT_DIR)/arch/$(BASE_DEV_NAME)/common
 
+EMB_COMMON_MOD_DIR := ${PREBOOT_ROOT_DIR}/emb_common_mod
+
 CFLAGS += -I$(SPECS_DIR)
 CFLAGS += -I$(SPECS_EMBEDDED_DIR)
 CFLAGS += -I$(SPECS_EMBEDDED_DIR)/$(BASE_DEV_NAME)
@@ -13,6 +15,10 @@
 CFLAGS += -Wall -Werror
 CFLAGS += -g -c
 CFLAGS += -I $(PREBOOT_ROOT_DIR)/include/ -I $(PREBOOT_ROOT_DIR)/ -I $(PREBOOT_ROOT_DIR)/lib/ -I $(PREBOOT_ROOT_DIR)/lib/common/ -I $(HL_DEV_ARCH)/
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/include
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib/common
+CFLAGS += -I $(EMB_COMMON_MOD_DIR)/lib/preboot
 
 CFLAGS += -Os -fdata-sections -ffunction-sections
 
@@ -28,6 +34,8 @@
 PLDM_BLOADER_SRCS_C += $(shell ls $(HL_DEV_ARCH_COMMON)/common.c)
 PLDM_BLOADER_SRCS_C += $(shell ls $(HL_DEV_ARCH)/common.c)
 PLDM_BLOADER_SRCS_C += $(shell ls $(HL_DEV_ARCH)/$(HL_TARGET).c)
+PLDM_BLOADER_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/lib/common/lib_common.c)
+PLDM_BLOADER_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/lib/common/pci.c)
 PLDM_BLOADER_APP_S := $(shell ls crt0_rom.S)
 
 include $(HL_DEV_ARCH)/Makefile
diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index f0a86b9..c57c82f 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -56,6 +56,7 @@
 #include <hl_boot_desc.h>
 #include <boot_info.h>
 #include <hl_sec_if.h>
+#include <hl_fit_if.h>
 
 #define PRODUCT_ID_STR_SIZE			0x8
 
@@ -250,7 +251,7 @@
 	image_entry();
 }
 
-void preboot_start_image(u64 startaddr, bool valid_arc_img)
+void preboot_start_image(u64 startaddr, bool valid_arc_img, __unused enum fit_core_id core_id)
 {
 	void (*image_entry)(void);
 	u32 lsb, msb;
diff --git a/arch/gaudi2/arm/common.c b/arch/gaudi2/arm/common.c
index 1927800..aeff2ec 100644
--- a/arch/gaudi2/arm/common.c
+++ b/arch/gaudi2/arm/common.c
@@ -44,6 +44,7 @@
 #include <hl.h>
 #include <address_space.h>
 #include <efuse.h>
+#include <hl_fit_if.h>
 
 
 #define ARC0_ICCM_BASE_ADDR		0x1000007FF7000000ull
@@ -193,7 +194,7 @@
 	init_arc();
 }
 
-void preboot_start_image(u64 startaddr, bool valid_arc_img)
+void preboot_start_image(u64 startaddr, bool valid_arc_img, __unused enum fit_core_id core_id)
 {
 	/*
 	 * Set ARC0 and ARC1 in running state. Don't run
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index e178930..21a1e96 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -755,14 +755,6 @@
 		hl_dev_pci_pll_wa();
 }
 
-void pci_dev_specific_mem_enabled(void)
-{
-	unsigned int rst_src = hl_rst_src_get();
-
-	if ((rst_src == HL_COLD_RST) || (rst_src == HL_PRSTN_RST))
-		hl_pci_iatu_init();
-}
-
 #define MSIX_PBA_OFFSET_REG_VAL	0x2002
 
 void set_pci_msix_pba_offset_val(void)
@@ -1216,11 +1208,6 @@
 	return boot_strap & PSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L_BTL_EN_MASK;
 }
 
-u32 dev_pci_dbi_cfg_offset_get(void)
-{
-	return (u32)mmPCIE_DBI_BASE & 0xFFFFFFF;
-}
-
 int preboot_set_sram_sf_access(bool is_enable)
 {
 	/*
@@ -1254,6 +1241,7 @@
 
 void device_init_pci(void)
 {
+	pci.dbi_base_addr = mmPCIE_DBI_BASE & 0xFFFFFFF;
 	pci.dbi_bar0_reg = mmPCIE_DBI_BAR0_REG;
 	pci.dbi_bar2_reg = mmPCIE_DBI_BAR2_REG;
 	pci.dbi_bar4_reg = mmPCIE_DBI_BAR4_REG;
diff --git a/arch/gaudi3/arc/Makefile b/arch/gaudi3/arc/Makefile
index ba3f472..17f5780 100644
--- a/arch/gaudi3/arc/Makefile
+++ b/arch/gaudi3/arc/Makefile
@@ -14,15 +14,26 @@
 endif
 
 PPBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH)/mbist.c)
+PPBOOT_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/$(HL_TARGET).c)
+PPBOOT_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/common.c)
+
+PLDM_BLOADER_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/$(HL_TARGET).c)
+PLDM_BLOADER_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/common.c)
+
+PIF_ARC_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/$(HL_TARGET).c)
+PIF_ARC_SRCS_C += $(shell ls $(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/common.c)
 
 EROM_PCIE_DEPS += \
 	$(PREBOOT_ROOT_DIR)/lib/common/boot_desc.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/efuse.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/string.o \
-	$(HL_DEV_ARCH_COMMON)/pci.o
+	$(HL_DEV_ARCH_COMMON)/pci.o \
+	$(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/$(HL_TARGET).o \
+	$(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/common.o
 
 EROM_RECOVERY_DEPS += \
-	$(HL_DEV_ARCH_COMMON)/pci.o \
-	$(HL_DEV_ARCH_COMMON)/fit.o
+	$(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/$(HL_TARGET).o \
+	$(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/common.o \
+	$(EMB_COMMON_MOD_DIR)/arch/$(BASE_DEV_NAME)/pci.o
 
 image_validation: image_validation_common
diff --git a/arch/gaudi3/arc/asic.c b/arch/gaudi3/arc/asic.c
index d6a1909..635c609 100644
--- a/arch/gaudi3/arc/asic.c
+++ b/arch/gaudi3/arc/asic.c
@@ -11,7 +11,6 @@
 #include <pcie_phy_regs.h>
 #include <pll_ctrl_regs.h>
 #include <gaudi3_reg_map.h>
-#include <common/pci_ids.h>
 
 #include <bitops.h>
 #include <baremetal.h>
@@ -239,11 +238,6 @@
 	preboot_puts("bist done for 48MB SRAM/CACHE\n");
 }
 
-u32 hl_skip_pcie_init(void)
-{
-	return hl_is_debug_mode(HL_DEBUG_HW_PCIE);
-}
-
 u64 hl_pll_timestamp_freq_get(void)
 {
 	/* As per Gaudi3 Frequencies.xlsx, PCI PLL Div[3] drives CPU ts */
@@ -289,16 +283,6 @@
 	regmap_writel(0xFFFFFFFF, D0_PCIE_DBI_SIG_REG(mmPCIE_DBI_BAR4_REG));
 	regmap_writel(0x0000001F, D0_PCIE_DBI_SIG_REG(mmPCIE_DBI_BAR5_REG));
 	regmap_writel(0x00000000, D0_PCIE_AUX_REG(mmPCIE_AUX_DBI));
-}
-
-u32 preboot_device_id_get(void)
-{
-	u32 dev_id = PCI_IDS_GAUDI3;
-
-	if (!hl_is_dual_die_sku())
-		dev_id = PCI_IDS_GAUDI3_SINGLE_DIE;
-
-	return dev_id;
 }
 
 #define PCIE_EXT_LD_TIMEOUT_US	10000
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index d24b9a8..b5ef9ce 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -12,12 +12,10 @@
 #include <qspi_regs.h>
 #include <pcie_aux_regs.h>
 #include <pcie_aux_masks.h>
-#include <pcie_core_regs.h>
 #include <pcie_dbi_regs.h>
 #include <psoc_reset_conf_regs.h>
 #include <efuse_regs.h>
 #include <uart_regs.h>
-#include <timestamp_regs.h>
 #include <i2c_regs.h>
 #include <int_agg_psoc_i2c_slv_comb_regs.h>
 #include <parc_local_cfg_regs.h>
@@ -145,11 +143,6 @@
 int hl_boot_info_set(u32 type, void *data, u32 size)
 {
 	return boot_info_set_no_merge(type, data, size);
-}
-
-u32 hl_pll_timestamp_base_get(void)
-{
-	return mmD0_PSOC_TIMESTAMP_BASE;
 }
 
 unsigned int get_cs_pad_sel_flash(void)
@@ -481,66 +474,9 @@
 	image_entry();
 }
 
-__weak bool hl_dev_spec_debug_mode(enum hl_debug_hw_func func)
-{
-	u32 boot_strap = regmap_readl(D0_PSOC_BOOT_CONF_REG(mmPSOC_BOOT_CONF_BOOT_STRAP_PINS_L));
-	u32 strap_mask;
-
-	switch (func) {
-	case HL_DEBUG_HW_PCIE:
-		strap_mask = BIT(28); /* BTL_DISABLE */
-		break;
-	case HL_DEBUG_HW_DRAM:
-		strap_mask = BIT(31); /* RSVD[1] */
-		break;
-	case HL_DEBUG_HW_I2C:
-		strap_mask = BIT(30); /* RSVD[0] */
-		break;
-	default:
-		preboot_puts_hex("Unknown H/W function ", func, false);
-		return 0;
-	}
-
-	return boot_strap & strap_mask;
-}
-
-bool hl_is_debug_mode(enum hl_debug_hw_func func)
-{
-	return hl_dev_spec_debug_mode(func);
-}
-
 u16 hl_get_boot_info_ver(void)
 {
 	return HL_BOOT_INFO_VERSION;
-}
-
-void preboot_copy_image(u64 dst, const void *buf, int len, const char *arch)
-{
-	u32 arc_dst = (u32)arcboot_rom_get_arc_addr(dst);
-
-	memcpy((void *)arc_dst, buf, len);
-}
-
-int preboot_decompress_image(u64 dst, const void *buf, int len,
-			     const char *arch, const char *compression)
-{
-	preboot_puts("Compression is not supported\n");
-	return 1;
-}
-
-void preboot_start_image(u64 startaddr, bool valid_arc_img)
-{
-	void (*image_entry)(void);
-	u32 arc1_addr = (u32)arcboot_rom_get_arc_addr(startaddr);
-
-	/* TODO: add start for ARC0 and ARC2 */
-
-	/*
-	 * Jump to __reset of zephyr. After image_entry, control is not
-	 * expected to return here and continue.
-	 */
-	image_entry = (void *)(arc1_addr);
-	image_entry();
 }
 
 uint32_t hl_get_die_id(void)
@@ -550,23 +486,6 @@
 	 * depending on the die location
 	 */
 	return regmap_readl(mmD0_PARC_LOCAL_CFG_BASE + mmPARC_LOCAL_CFG_DIE_ID);
-}
-
-#define EFUSE_SKU_OFFSET	58
-#define EFUSE_SKU_SHIFT		26
-
-bool hl_is_dual_die_sku(void)
-{
-	u32 val;
-	int err;
-
-	err = preboot_fuse_read(EFUSE_SKU_OFFSET, &val);
-	if (err)
-		return false;
-
-	val = (val >> EFUSE_SKU_SHIFT) & 0x1;
-
-	return !val;
 }
 
 bool hl_is_sim_mode(void)
@@ -609,27 +528,6 @@
 	uart.uart_ctrl_reg = D0_PSOC_UART_0_REG(mmUART_CTRL);
 	uart.uart_bauddiv_reg = D0_PSOC_UART_0_REG(mmUART_BAUDDIV);
 	uart.regs_initialized = true;
-}
-
-void device_init_pci(void)
-{
-	/* DIE 0 */
-	pci.dbi_bar0_reg = D0_PCIE_DBI_SIG_REG(mmPCIE_DBI_BAR0_REG);
-	pci.dbi_bar2_reg = D0_PCIE_DBI_SIG_REG(mmPCIE_DBI_BAR2_REG);
-	pci.dbi_bar4_reg = D0_PCIE_DBI_SIG_REG(mmPCIE_DBI_BAR4_REG);
-	pci.dbi_status_command_reg = D0_PCIE_DBI_SIG_REG(mmPCIE_DBI_STATUS_COMMAND_REG);
-	pci.dbi_uncorr_aer_mask = D0_PCIE_DBI_SIG_REG(mmPCIE_DBI_UNCORR_ERR_MASK_OFF);
-	pci.aux_dbi_reg = D0_PCIE_AUX_REG(mmPCIE_AUX_DBI);
-	pci.aux_mem_space_en_reg = D0_PCIE_AUX_REG(mmPCIE_AUX_MEM_SPACE_EN);
-	pci.aux_flr_ctrl_reg = D0_PCIE_AUX_REG(mmPCIE_AUX_FLR_CTRL);
-	pci.aux_ltssm_en_reg = D0_PCIE_AUX_REG(mmPCIE_AUX_LTSSM_EN);
-	pci.aux_pf_flr_active_reg = D0_PCIE_AUX_REG(mmPCIE_AUX_PF_FLR_ACTIVE);
-	pci.aux_flr_int_reg = D0_PCIE_AUX_REG(mmPCIE_AUX_FLR_INT);
-	pci.aux_pf_flr_done_reg = D0_PCIE_AUX_REG(mmPCIE_AUX_PF_FLR_DONE);
-	pci.aux_rdlh_link_up_reg = D0_PCIE_AUX_REG(mmPCIE_AUX_RDLH_LINK_UP);
-	pci.aux_smlh_link_up_reg = D0_PCIE_AUX_REG(mmPCIE_AUX_SMLH_LINK_UP);
-	pci.core_hot_rst_reg = D0_PCIE_CORE_REG(mmPCIE_CORE_HOT_RST);
-	pci.bmu_base_addr = mmD0_PCIE_BMU_BASE;
 }
 
 void device_init_i2c_s(void)
@@ -710,35 +608,6 @@
 void lkd_dev_specific_setup(void)
 {
 	cpu_boot_dev_sts_set(CPU_BOOT_DEV_STS_FW_HARD_RST_EN);
-}
-
-bool preboot_is_valid_address_range(u64 addr, int len)
-{
-	if (addr >= SP_SRAM_BASE &&
-	    addr + len < SP_SRAM_BASE + HL_SP_SRAM_SIZE)
-		return true;
-
-	if (addr >= HL_SP_SRAM_BASE &&
-	    addr + len < HL_SP_SRAM_BASE + HL_SP_SRAM_SIZE)
-		return true;
-
-	if (addr >= ARC0_ICCM_FULL_BASE &&
-	    addr + len < ARC0_ICCM_FULL_BASE + ARC0_MEM_SIZE)
-		return true;
-
-	if (addr >= ARC1_ICCM_FULL_BASE &&
-	    addr + len < ARC1_ICCM_FULL_BASE + ARC1_MEM_SIZE)
-		return true;
-
-	if (addr >= HL_SRAM_BASE_FULL && addr + len <= HL_SRAM_BASE_FULL + SDRAM_SIZE)
-		return true;
-
-	if (addr >= SDRAM_BASE && addr + len < SDRAM_BASE + SDRAM_SIZE)
-		return true;
-
-	preboot_puts_hex64("ERROR: invalid address range addr = ", addr, true);
-
-	return false;
 }
 
 static u32 ppboot_get_arc_addr(u64 addr)
diff --git a/arch/gaudi3/arc/common.h b/arch/gaudi3/arc/common.h
index 7b7a17c..5b4878b 100644
--- a/arch/gaudi3/arc/common.h
+++ b/arch/gaudi3/arc/common.h
@@ -35,6 +35,7 @@
 
 #define ARC0_MEM_SIZE			(ARC0_ICCM_SIZE + ARC0_DCCM_SIZE)
 #define ARC1_MEM_SIZE			(ARC1_ICCM_SIZE + ARC1_DCCM_SIZE)
+#define ARC2_MEM_SIZE			(ARC2_ICCM_SIZE + ARC2_DCCM_SIZE)
 
 #define ARC0_ICCM_FULL_BASE		0x300007FD3000000ull
 #define ARC0_ICCM_SIZE			SZ_128K
@@ -46,6 +47,12 @@
 #define ARC1_DCCM_SIZE			SZ_64K
 #define ARC2_ICCM_FULL_BASE		0x300007FD30c0000ull
 #define ARC2_ICCM_SIZE			SZ_128K
+#define ARC2_DCCM_FULL_BASE		0x300007FD30E0000ull
+#define ARC2_DCCM_SIZE			SZ_32K
+
+#define ARC1_OFFLOAD_DCCM_FULL_BASE    0x0300007FDB0E0000ull
+#define ARC1_OFFLOAD_DCCM_SIZE         SZ_32K
+
 
 #define SRAM_MEM_REGION_BASE		0x10000000
 #define FW_MEM_REGION_BASE		0x20000000
diff --git a/arch/gaudi3/arc/dev_ext_rom.h b/arch/gaudi3/arc/dev_ext_rom.h
index c29a668..3b46d30 100644
--- a/arch/gaudi3/arc/dev_ext_rom.h
+++ b/arch/gaudi3/arc/dev_ext_rom.h
@@ -8,7 +8,7 @@
 
 #include <gaudi3/gaudi3_mem_addr_dt.h>
 
-#define EXT_ROM_TEXT_LENGTH		(64*1024)
+#define EXT_ROM_TEXT_LENGTH		(70*1024)
 /* No header bytes. Hence the text base is same as erom start address */
 #define EXT_ROM_PCIE_TEXT_BASE		(EROM_START_OFFSET)
 #define EXT_ROM_RECOVERY_TEXT_BASE	(EROM_START_OFFSET)
diff --git a/arch/gaudi3/arc/hl.h b/arch/gaudi3/arc/hl.h
index 98599bb..c63dc72 100644
--- a/arch/gaudi3/arc/hl.h
+++ b/arch/gaudi3/arc/hl.h
@@ -69,5 +69,6 @@
 bool ppboot_is_valid_img_desc_ex(void *desc);
 bool is_spi_gw_enabled(void);
 u32 gaudi3_count_set_bit(u32 n);
+void gaudi3_clean_hot_reset_intr(void);
 
 #endif /* ARCH_GAUDI3_ARC_HL_H_ */
diff --git a/arch/gaudi3/arc/onix.c b/arch/gaudi3/arc/onix.c
index ab603cf..4e8a256 100644
--- a/arch/gaudi3/arc/onix.c
+++ b/arch/gaudi3/arc/onix.c
@@ -7,7 +7,6 @@
 #include <global_conf_regs.h>
 #include <arc_regs.h>
 #include <gaudi3_reg_map.h>
-#include <common/pci_ids.h>
 
 #include <gaudi3/address_space.h>
 #include <baremetal.h>
@@ -15,31 +14,11 @@
 
 #define SYS_CLOCK_FREQUENCY	35000000ULL
 
-u64 hl_pll_timestamp_freq_get(void)
-{
-	return SYS_CLOCK_FREQUENCY;
-}
-
-u32 hl_skip_pcie_init(void)
-{
-	return 1;
-}
-
 void hl_uart_init(void)
 {
 	regmap_writel(1, D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_PAD_SEL_48));	/* UART0 RXD */
 	regmap_writel(1, D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_PAD_SEL_49));	/* UART0 TXD */
 	preboot_uart_init(SYS_CLOCK_FREQUENCY / (HL_UART_BAUDRATE / 2));
-}
-
-u32 hl_spi_baudr_get(void)
-{
-	u32 factor = SYS_CLOCK_FREQUENCY / HL_SPI_FLASH_MAX_FREQ;
-
-	/* baud rate doesn't support odd number and must not be 0 */
-	factor = (factor < 2) ? 2 : factor;
-
-	return (factor + 1) & 0xFFFE;
 }
 
 /*
@@ -68,11 +47,6 @@
 	}
 
 	return addr;
-}
-
-u32 preboot_device_id_get(void)
-{
-	return PCI_IDS_GAUDI3_FPGA;
 }
 
 /* plls are not implemented on onix platform */
diff --git a/arch/gaudi3/arc/pldm.c b/arch/gaudi3/arc/pldm.c
index 349d599..a1f5a2f 100644
--- a/arch/gaudi3/arc/pldm.c
+++ b/arch/gaudi3/arc/pldm.c
@@ -7,7 +7,6 @@
 #include <global_conf_regs.h>
 #include <global_conf_masks.h>
 #include <arc_regs.h>
-#include <common/pci_ids.h>
 #include <stddef.h>
 #include <bitops.h>
 #include <baremetal.h>
@@ -17,10 +16,10 @@
 #include <boot_desc.h>
 #include <efuse.h>
 #include "hl.h"
+#include <fit.h>
 
-#define SYS_CLOCK_FREQUENCY	250000000ULL /* PCI PLL DIV3 SIMULATION CLK */
+#define UART_CLOCK_FREQUENCY	250000000ULL /* PCI PLL DIV3 SIMULATION CLK */
 
-#define UART_CLOCK_FREQUENCY	SYS_CLOCK_FREQUENCY
 #define PLDM_UART_BAUDRATE	300
 
 /*
@@ -73,31 +72,11 @@
 	},
 };
 
-u64 hl_pll_timestamp_freq_get(void)
-{
-	return SYS_CLOCK_FREQUENCY;
-}
-
-u32 hl_skip_pcie_init(void)
-{
-	return 0;
-}
-
 void hl_uart_init(void)
 {
 	regmap_writel(1, D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_PAD_SEL_48)); /* UART0 RXD */
 	regmap_writel(1, D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_PAD_SEL_49)); /* UART0 TXD */
 	preboot_uart_init(UART_CLOCK_FREQUENCY / PLDM_UART_BAUDRATE);
-}
-
-u32 hl_spi_baudr_get(void)
-{
-	u32 factor = SYS_CLOCK_FREQUENCY / HL_SPI_FLASH_MAX_FREQ;
-
-	/* baud rate doesn't support odd number and must not be 0 */
-	factor = (factor < 2) ? 2 : factor;
-
-	return (factor + 1) & 0xFFFE;
 }
 
 #define SP_SRAM_INIT_TIMEOUT_US		(5 * USEC_PER_SEC)
@@ -155,16 +134,6 @@
 bool hl_dev_is_valid_img_desc(struct boot_desc *desc)
 {
 	return true;
-}
-
-u32 preboot_device_id_get(void)
-{
-	u32 dev_id = PCI_IDS_GAUDI3;
-
-	if (!hl_is_dual_die_sku())
-		dev_id = PCI_IDS_GAUDI3_SINGLE_DIE;
-
-	return dev_id;
 }
 
 #define PCI_AUX_DIV_ID	2
@@ -245,5 +214,5 @@
 void btldr_exec_app(void)
 {
 	preboot_puts("pldm-bloader resigned\n");
-	preboot_start_image(APP_RST_VECTOR_ADDR, true);
+	preboot_start_image(APP_RST_VECTOR_ADDR, true, FIT_CORE_ID_D0_PSOC_ARC1);
 }
diff --git a/arch/gaudi3/common/common.c b/arch/gaudi3/common/common.c
index f171861..2bd33b7 100644
--- a/arch/gaudi3/common/common.c
+++ b/arch/gaudi3/common/common.c
@@ -120,14 +120,6 @@
 		D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_COLD_RST_FLOPS_2));
 }
 
-bool preboot_sram_is_valid_address_range(u64 addr, int len)
-{
-	if (addr >= SDRAM_BASE && addr + len <= SDRAM_BASE + SRAM_SIZE)
-		return !preboot_is_overlap_address_range(addr, len);
-
-	return false;
-}
-
 bool tpm_is_supported(void) { return true; }
 
 void spi_tpm_base_get(unsigned int *regs)
diff --git a/arch/gaudi3/common/fit.c b/arch/gaudi3/common/fit.c
deleted file mode 100644
index 4b72adc..0000000
--- a/arch/gaudi3/common/fit.c
+++ /dev/null
@@ -1,26 +0,0 @@
-// SPDX-License-Identifier: GPL-2.0+
-/*
- * Copyright (C) 2022 HabanaLabs Ltd.
- */
-
-#include <stddef.h>
-#include <stdint.h>
-#include <stdbool.h>
-#include <types.h>
-#include <fit.h>
-#include <uart.h>
-
-bool fit_is_core_id_supported(void)
-{
-	return true;
-}
-
-int fit_process_core_id(enum fit_core_id core_id)
-{
-	/* recovery currently only support fit load for master psoc arc1 */
-	if (core_id == FIT_CORE_ID_D0_PSOC_ARC1)
-		return 0;
-
-	preboot_puts_hex_raw("Unsupported core id: ", core_id, false);
-	return 1;
-}
diff --git a/arch/gaudi3/common/pci.c b/arch/gaudi3/common/pci.c
index 97f186b..c97cb57 100644
--- a/arch/gaudi3/common/pci.c
+++ b/arch/gaudi3/common/pci.c
@@ -8,7 +8,6 @@
 #include <baremetal.h>
 #include <gaudi3_blocks_linux_driver.h>
 #include <gaudi3_reg_map.h>
-#include <gaudi3_fw_if.h>
 #include <global_conf_regs.h>
 #include <psoc_boot_conf_regs.h>
 #include <psoc_boot_conf_masks.h>
@@ -17,234 +16,10 @@
 #include <pcie_phy_masks.h>
 #include <pcie_phy_regs.h>
 #include <pcie_wrap_regs.h>
-#include <hl_pci_if.h>
-#include <hl_boot_if.h>
-#include <address_space.h>
-#include <boot_if.h>
-#include <bitops.h>
-#include <device.h>
-#include <uart.h>
 #include <pci.h>
 #include <hl.h>
 #include <pll.h>
 #include <arcboot_export_dev.h>
-
-static bool pcie_initialized;
-
-/* TODO: Move this functions to emb common modules */
-bool is_pcie_initialized(void)
-{
-	unsigned int prstn, smlh, rdlh;
-
-	if (pcie_initialized)
-		return true;
-
-	prstn = regmap_readl(global_conf.prstn_reg);
-	smlh = regmap_readl(pci.aux_rdlh_link_up_reg);
-	rdlh = regmap_readl(pci.aux_smlh_link_up_reg);
-
-	if (prstn && smlh && rdlh) {
-		pcie_initialized = true;
-		return true;
-	}
-
-	return false;
-}
-
-u32 dev_pci_dbi_cfg_offset_get(void)
-{
-	return mmD0_PCIE_DBI_SIG_BASE;
-}
-
-/* TODO: Move this functions to emb common modules */
-void pci_dev_specific_setup(void)
-{
-	cpu_boot_dev_sts_set(CPU_BOOT_DEV_STS_FW_IATU_CONF_EN);
-}
-
-void pci_dev_flr_enable(bool enable)
-{
-	/* By default FLR RST handled by HW */
-}
-
-void pci_dev_specific_flr(void)
-{
-	regmap_writel(PCIE_AUX_FLR_INT_IND_M, D0_PCIE_AUX_REG(mmPCIE_AUX_FLR_INT));
-}
-
-void pci_dev_specific_pll_init(void)
-{
-	/* Nothing need to implemented for gaudi3 */
-}
-
-void pci_dev_specific_mem_enabled(void)
-{
-	hl_pci_bmu_init();
-	hl_pci_iatu_init();
-}
-
-struct hl_pci_inbound_region gaudi3_pci_inbound_regions[] = PCI_IATU_INBOUND_REGIONS_DATA;
-
-/* TODO: Move this functions to emb common modules */
-void hl_pci_iatu_init(void)
-{
-	int i;
-	struct hl_pci_outbound_region outbound_region;
-
-	for (i = 0; i < ARRAY_SIZE(gaudi3_pci_inbound_regions); ++i)
-		hl_pci_set_inbound_region(i, &gaudi3_pci_inbound_regions[i]);
-
-	/* Outbound Region 0 - Point to Host */
-	outbound_region.addr = HOST_PHYS_BASE_0;
-	outbound_region.size = HOST_PHYS_SIZE_0;
-	hl_pci_set_outbound_region(&outbound_region);
-}
-
-struct hl_bmu_region gaudi3_pci_bmu_regions[] = PCI_BMU_REGIONS_DATA;
-
-/* TODO: Move this functions to emb common modules */
-void hl_pci_bmu_init(void)
-{
-	int err, i;
-
-	hl_pci_bmu_set_cfg_base(CFG_PHYS_BASE);
-
-	for (i = 0; i < ARRAY_SIZE(gaudi3_pci_bmu_regions); ++i) {
-		err = hl_pci_set_bmu_region(&gaudi3_pci_bmu_regions[i]);
-		if (err)
-			preboot_puts_hex_raw("error initializing bmu region ", i, false);
-	}
-
-	hl_pci_bmu_enable(true);
-}
-
-#define BMU_CFG_SRC_START_OFF		0x8
-#define BMU_CFG_SRC_END_OFF			0x48
-#define BMU_CFG_DST_OFFS_R0_OFF		0x88
-#define BMU_CFG_DST_OFFS_R1_OFF		0xC8
-#define BMU_CFG_AWUSER_R0_OFF		0x108
-#define BMU_CFG_AWUSER_R1_OFF		0x148
-#define BMU_CFG_ARUSER_R0_OFF		0x1C8
-#define BMU_CFG_ARUSER_R1_OFF		0x208
-#define BMU_CFG_AXI_AW_PROT_OFF		0x288
-#define BMU_CFG_AXI_AW_CACHE_OFF	0x2A8
-#define BMU_CFG_AXI_AR_PROT_OFF		0x2C8
-#define BMU_CFG_AXI_AR_CACHE_OFF	0x2E8
-#define BMU_CFG_ADDR_SHIFT			2
-#define BMU_CFG_OFF_MASK			0xFFFFFFFFFCull /* GENMASK(39, 2) */
-
-static void hl_pci_set_bmu_cfg_region(struct hl_bmu_region *region)
-{
-	u32 base_addr = pci.bmu_base_addr + (region->id * SZ_4);
-	u32 start_addr = (u32)(region->src_start >> BMU_CFG_ADDR_SHIFT);
-	u32 end_addr = (u32)(region->src_end >> BMU_CFG_ADDR_SHIFT);
-	u64 dst_addr = region->dst_base - region->src_start;
-	u64 dst_off = (dst_addr & BMU_CFG_OFF_MASK) >> BMU_CFG_ADDR_SHIFT;
-
-	regmap_writel(start_addr, base_addr + BMU_CFG_SRC_START_OFF);
-	regmap_writel(end_addr, base_addr + BMU_CFG_SRC_END_OFF);
-	regmap_writel((u32)(dst_off), base_addr + BMU_CFG_DST_OFFS_R0_OFF);
-	regmap_writel((u32)(dst_off >> 32), base_addr + BMU_CFG_DST_OFFS_R1_OFF);
-
-	regmap_writel((u32)(region->perms.w_user), base_addr + BMU_CFG_AWUSER_R0_OFF);
-	regmap_writel((u32)(region->perms.w_user >> 32), base_addr + BMU_CFG_AWUSER_R1_OFF);
-	regmap_writel((u32)(region->perms.r_user), base_addr + BMU_CFG_ARUSER_R0_OFF);
-	regmap_writel((u32)(region->perms.r_user >> 32), base_addr + BMU_CFG_ARUSER_R1_OFF);
-	regmap_writel(region->perms.w_prot, base_addr + BMU_CFG_AXI_AW_PROT_OFF);
-	regmap_writel(region->perms.r_prot, base_addr + BMU_CFG_AXI_AR_PROT_OFF);
-	regmap_writel(region->perms.w_cache, base_addr + BMU_CFG_AXI_AW_CACHE_OFF);
-	regmap_writel(region->perms.r_cache, base_addr + BMU_CFG_AXI_AR_CACHE_OFF);
-}
-
-#define BMU_MEM_SRC_START_OFF		0x500
-#define BMU_MEM_SRC_END_OFF			0x520
-#define BMU_MEM_DST_OFFS_OFF		0x540
-#define BMU_MEM_AWUSER_R0_OFF		0x560
-#define BMU_MEM_AWUSER_R1_OFF		0x580
-#define BMU_MEM_ARUSER_R0_OFF		0x5C0
-#define BMU_MEM_ARUSER_R1_OFF		0x5E0
-#define BMU_MEM_AXI_AW_PROT_OFF		0x620
-#define BMU_MEM_AXI_AW_CACHE_OFF	0x630
-#define BMU_MEM_AXI_AR_PROT_OFF		0x640
-#define BMU_MEM_AXI_AR_CACHE_OFF	0x650
-#define BMU_MEM_DST_BASE_OFF		0x660
-#define BMU_MEM_ADDR_SHIFT			20
-#define BMU_MEM_DEST_BASE_SHIFT		48
-#define BMU_MEM_SRC_MASK		0x3FFFF00000ull	  /* GENMASK(37, 20) */
-#define BMU_MEM_DEST_MASK		0xFFFFFFF00000ull /* GENMASK(47, 20) */
-
-static void hl_pci_set_bmu_mem_region(struct hl_bmu_region *region)
-{
-	u32 base_addr = pci.bmu_base_addr + (region->id * SZ_4);
-	u32 start_addr = (u32)((region->src_start & BMU_MEM_SRC_MASK) >> BMU_MEM_ADDR_SHIFT);
-	u32 end_addr = (u32)((region->src_end & BMU_MEM_SRC_MASK) >> BMU_MEM_ADDR_SHIFT);
-	u32 mem_base = (u32)(region->dst_base >> BMU_MEM_DEST_BASE_SHIFT);
-	u64 dst_addr = region->dst_base - region->src_start;
-	u32 dst_off = (u32)((dst_addr & BMU_MEM_DEST_MASK) >> BMU_MEM_ADDR_SHIFT);
-
-	regmap_writel(start_addr, base_addr + BMU_MEM_SRC_START_OFF);
-	regmap_writel(end_addr, base_addr + BMU_MEM_SRC_END_OFF);
-	regmap_writel(mem_base, base_addr + BMU_MEM_DST_BASE_OFF);
-	regmap_writel(dst_off, base_addr + BMU_MEM_DST_OFFS_OFF);
-
-	regmap_writel((u32)(region->perms.w_user), base_addr + BMU_MEM_AWUSER_R0_OFF);
-	regmap_writel((u32)(region->perms.w_user >> 32), base_addr + BMU_MEM_AWUSER_R1_OFF);
-	regmap_writel((u32)(region->perms.r_user), base_addr + BMU_MEM_ARUSER_R0_OFF);
-	regmap_writel((u32)(region->perms.r_user >> 32), base_addr + BMU_MEM_ARUSER_R1_OFF);
-	regmap_writel(region->perms.w_prot, base_addr + BMU_MEM_AXI_AW_PROT_OFF);
-	regmap_writel(region->perms.r_prot, base_addr + BMU_MEM_AXI_AR_PROT_OFF);
-	regmap_writel(region->perms.w_cache, base_addr + BMU_MEM_AXI_AW_CACHE_OFF);
-	regmap_writel(region->perms.r_cache, base_addr + BMU_MEM_AXI_AR_CACHE_OFF);
-}
-
-#define BMU_REGION_EN_OFF	0x4
-#define BMU_NUM_CFG_REGIONS	8
-
-static void hl_pci_bmu_enable_region(enum hl_bmu_region_id id, enum hl_bmu_region_type type)
-{
-	u32 reg = regmap_readl(pci.bmu_base_addr + BMU_REGION_EN_OFF);
-	u32 reg_mask = BIT(id);
-
-	reg_mask = (type == BMU_REGION_TYPE_MEM) ? (reg_mask << BMU_NUM_CFG_REGIONS) : reg_mask;
-	reg |= reg_mask;
-	regmap_writel(reg, pci.bmu_base_addr + BMU_REGION_EN_OFF);
-}
-
-int hl_pci_set_bmu_region(struct hl_bmu_region *region)
-{
-	if (region->type == BMU_REGION_TYPE_MEM) {
-		if (region->id > BMU_MEM_REGION_MAX)
-			return -1;
-
-		hl_pci_set_bmu_mem_region(region);
-	} else {
-		if (region->id > BMU_CFG_REGION_MAX)
-			return -1;
-
-		hl_pci_set_bmu_cfg_region(region);
-	}
-
-	hl_pci_bmu_enable_region(region->id, region->type);
-
-	return 0;
-}
-
-#define BMU_CFG_BASE_SHIFT	40
-#define BMU_CFG_DST_BASE_OFFSET	0x308
-
-void hl_pci_bmu_set_cfg_base(u64 cfg_base_addr)
-{
-	u32 cfg_dst = (u32)(cfg_base_addr >> BMU_CFG_BASE_SHIFT);
-
-	regmap_writel(cfg_dst, pci.bmu_base_addr + BMU_CFG_DST_BASE_OFFSET);
-}
-
-void hl_pci_bmu_enable(bool enable)
-{
-	u32 val = enable ? 1 : 0;
-
-	regmap_writel(val, pci.bmu_base_addr);
-}
 
 #define PHY_CR_PARA_CMD_DATA		0x2
 #define PHY_CR_PARA_CMD_ADDR		0x5149
diff --git a/include/baremetal.h b/include/baremetal.h
index df23f13..e256fbe 100644
--- a/include/baremetal.h
+++ b/include/baremetal.h
@@ -46,6 +46,7 @@
 #define FIT_IMG_SPSRAM_END		(HL_SP_SRAM_BASE + HL_SP_SRAM_SIZE)
 #define FIT_IMG_SPSRAM_BASE		(FIT_IMG_SPSRAM_END - \
 					 FIT_IMG_SPSRAM_MAX_SIZE)
+#define FIT_CONF_NAME			"recovery-conf"
 
 
 #ifndef __ASSEMBLY__
@@ -204,6 +205,7 @@
 bool preboot_host_gpio_asserted(void);
 void preboot_sp_sram_init(void);
 void preboot_cond_arc0_reset(void);
+void preboot_configure_arc(int id, uint64_t startaddr);
 
 void ppboot_cp_preboot(u64 loadaddr, u32 offset, u32 size, char *name);
 void image_start(u64 loadaddr);
@@ -246,14 +248,14 @@
 void preboot_ack_irq(void);
 const char *preboot_get_version(enum preboot_vers ver);
 void hl_ram_init(void);
+void hl_pcie_init(void);
 void hl_pci_iatu_init(void);
 void hl_pci_bmu_init(void);
 bool hl_pci_link_is_up(void);
 void hl_pcie_enable_access(void);
 void hl_pcie_disable_access(void);
-void hl_pcie_enable_mem_access(void);
-void hl_pcie_disable_mem_access(void);
 bool hl_is_pcie_link_up(void);
+void pci_init_wait_dbg(void);
 
 void preboot_finalize_fw_info(void);
 void preboot_finalize_nic_en(void);
@@ -265,7 +267,6 @@
 int preboot_verify_img_sig(const void *fit, size_t img_size,
 			   enum fit_sec_failure *res, u32 *key_idx);
 size_t preboot_get_img_sec_desc_size(void);
-void preboot_start_image(u64 startaddr, bool valid_arc_img);
 
 u32 preboot_error_read_and_clr(u32 *major, u32 *minor);
 bool preboot_error_inform(u32 major, u32 minor);
diff --git a/include/device.h b/include/device.h
index fcd32f9..65fedba 100644
--- a/include/device.h
+++ b/include/device.h
@@ -22,6 +22,7 @@
 };
 
 struct device_pci {
+	u32 dbi_base_addr;
 	u32 dbi_bar0_reg;
 	u32 dbi_bar2_reg;
 	u32 dbi_bar4_reg;
diff --git a/lib/common/cmd_module.h b/lib/common/cmd_module.h
index de590e8..e861fe9 100644
--- a/lib/common/cmd_module.h
+++ b/lib/common/cmd_module.h
@@ -84,7 +84,7 @@
 void pci_dev_specific_flr(void);
 void pci_dev_flr_enable(bool enable);
 void pci_dev_specific_pll_init(void);
-void pci_dev_specific_mem_enabled(void);
+void pci_dev_specific_post_link_up(void);
 
 /* FWUPD */
 void module_fwupd_setup(void);
@@ -94,6 +94,7 @@
 void module_fwupd_run_noret(void);
 void module_fwupd_pcie_disable(bool is_pcie_disabled);
 bool module_fwupd_is_disabled(void);
+void module_fwupd_store_pcie_access(void);
 
 /* I2C FWUPD */
 void module_i2c_fwupd_setup(void);
diff --git a/lib/common/lib_common.c b/lib/common/lib_common.c
deleted file mode 100644
index c8746a9..0000000
--- a/lib/common/lib_common.c
+++ /dev/null
@@ -1,11 +0,0 @@
-// SPDX-License-Identifier:     GPL-2.0+
-/*
- * Copyright (C) 2017-2022 HabanaLabs Ltd.
- * All Rights Reserved.
- *
- */
-
-#include <types.h>
-#include <device.h>
-
-struct device_global_conf global_conf;
diff --git a/lib/common/module_pci.c b/lib/common/module_pci.c
index 7e1c50d..df43db0 100644
--- a/lib/common/module_pci.c
+++ b/lib/common/module_pci.c
@@ -15,18 +15,6 @@
 #include "pll.h"
 #include <hl.h>
 
-struct pci_status {
-	u32 flr_reg_val;
-	u32 hot_rst_reg_val;
-	u32 smlh_link_reg_val;
-	u32 rdlh_link_reg_val;
-	bool link_is_up;
-	bool mem_enabled;
-	bool mem_enabled_prev;
-	bool pci_pll_initialized;
-	bool setup_done;
-};
-
 static struct pci_status pci_sts = {
 	.flr_reg_val = ~1,
 	.hot_rst_reg_val = ~1,
@@ -38,6 +26,7 @@
 	.setup_done = false,
 };
 
+__weak void pci_dev_specific_setup(void) {}
 __weak void pci_dev_specific_teardown(void) {}
 __weak void pci_dev_specific_disable(void) {}
 __weak void pci_dev_specific_enable(void) {}
@@ -66,24 +55,15 @@
 	regmap_writel(1, pci.aux_pf_flr_done_reg);
 }
 
+__weak bool is_pcie_initialized(void) { return true; }
 __weak void hl_dev_pci_pll_wa(void) {}
-
-__weak void module_pci_do_setup(void)
-{
-	if (!hl_is_debug_mode(HL_DEBUG_HW_PCIE)) {
-		pci_dev_specific_setup();
-		pci_dev_flr_enable(true);
-	}
-
-	pci_sts.setup_done = true;
-}
 
 void module_pci_setup(void)
 {
 	if (!is_pcie_initialized())
 		return;
 
-	module_pci_do_setup();
+	module_pci_do_common_setup(&pci_sts);
 }
 
 __weak u32 pci_dev_specific_hot_rst_check(void)
@@ -117,9 +97,6 @@
 	return NOT_FOUND;
 }
 
-__weak bool pci_dev_specific_pif_err_pending(void) { return false; }
-__weak void pci_dev_specific_pif_err_process(void) {}
-
 enum cmd_check_res module_pci_check(void)
 {
 	if (hl_is_debug_mode(HL_DEBUG_HW_PCIE))
@@ -136,7 +113,7 @@
 	}
 
 	if (!pci_sts.setup_done)
-		module_pci_do_setup();
+		module_pci_do_common_setup(&pci_sts);
 
 	return dev_specific_module_pci_check();
 }
@@ -153,6 +130,9 @@
 
 	pci_sts.setup_done = false;
 }
+
+
+__weak void pci_dev_specific_pll_init(void) { }
 
 __weak void pci_dev_specific_run(void)
 {
@@ -172,38 +152,7 @@
 
 void module_pci_run_fast(void)
 {
-	if (pci_sts.flr_reg_val == 1) {
-		pci_dev_specific_flr();
-		preboot_puts("FLR received\n");
-	}
-
-	if (pci_sts.hot_rst_reg_val == 1) {
-		pci_dev_specific_hot_rst();
-		preboot_puts("HOT RST received\n");
-	}
-
-	if ((pci_sts.rdlh_link_reg_val && pci_sts.smlh_link_reg_val) !=
-	    pci_sts.link_is_up) {
-		pci_sts.link_is_up = !pci_sts.link_is_up;
-		preboot_puts_raw("PCIe link is ");
-		if (pci_sts.link_is_up) {
-			preboot_puts("up\n");
-			pci_dev_specific_post_link_up();
-		} else {
-			preboot_puts("down\n");
-			pci_sts.mem_enabled = false;
-		}
-	}
-
-	if (pci_sts.mem_enabled_prev != pci_sts.mem_enabled) {
-		pci_sts.mem_enabled_prev = pci_sts.mem_enabled;
-		if (pci_sts.mem_enabled)
-			pci_dev_specific_mem_enabled();
-
-		preboot_puts_raw("PCIe mem_en changed to: ");
-		preboot_puts_raw(pci_sts.mem_enabled ? "true" : "false");
-		preboot_puts("\n");
-	}
+	module_pci_do_common_run(&pci_sts);
 }
 
 void module_pci_enable(void)
diff --git a/lib/common/pci.c b/lib/common/pci.c
deleted file mode 100644
index 38c1618..0000000
--- a/lib/common/pci.c
+++ /dev/null
@@ -1,259 +0,0 @@
-// SPDX-License-Identifier:     GPL-2.0+
-/*
- * Copyright (C) 2017-2021 HabanaLabs Ltd.
- * All Rights Reserved.
- *
- */
-
-#include <common/pci_ids.h>
-#include <bitops.h>
-#include <baremetal.h>
-#include <device.h>
-#include <sizes.h>
-#include "pci.h"
-#include "uart.h"
-#include "pll.h"
-
-#define IATU_BAR_NUM_OFFSET		8
-#define IATU_BAR_NUM_MASK		0x700
-#define IATU_NUM_MATCH_EN_MASK		BIT(19)
-#define IATU_MATCH_MODE_MASK		BIT(30)
-#define IATU_REGION_EN_MASK		BIT(31)
-
-#define IATU_REGION_SIZE		SZ_512
-#define IATU_REGION_INBOUND_OFF		0x100
-#define IATU_REGION_CTRL_1_OFF		0x0
-#define IATU_REGION_CTRL_2_OFF		0x4
-#define IATU_LWR_BASE_ADDR_OFF		0x8
-#define IATU_UPPER_BASE_ADDR_OFF	0xC
-#define IATU_LIMIT_ADDR_OFF		0x10
-#define IATU_LWR_TARGET_ADDR_OFF	0x14
-#define IATU_UPPER_TARGET_ADDR_OFF	0x18
-#define IATU_REGION_CTRL_3_OFF		0x1C
-#define IATU_UPPR_LIMIT_ADDR_OFF	0x20
-
-#define IATU_ACCESS_ENABLE_VAL		0x00300000
-#define IATU_ACCESS_DISABLE_VAL		0
-#define IATU_OUTBOUND_REGION_SIZE	SZ_8K
-#define IATU_OUTBOUND_UPPER_ADDR_VAL	0x08000000
-
-struct device_pci pci;
-
-__weak u32 dev_pci_dbi_cfg_offset_get(void)
-{
-	return 0;
-}
-
-__weak void hl_pcie_enable_access(void) {}
-__weak void hl_pcie_disable_access(void) {}
-
-/*
- * hl_pci_get_bar_phys_addr() - get physical address of a bar
- * @bar_id: bar number
- * @bar_phys_addr: returned bar physical address
- *
- * the physical address is the address allocated by the host to each bar.
- */
-static int hl_pci_get_bar_phys_addr(enum hl_pci_bar_id bar_id, u64 *bar_phys_addr)
-{
-	u32 bar_reg, bar_addr_low, bar_addr_high;
-	u32 aux_dbi_reg;
-
-	switch (bar_id) {
-	case PCI_BAR_ID_0:
-		bar_reg = pci.dbi_bar0_reg;
-		break;
-	case PCI_BAR_ID_2:
-		bar_reg = pci.dbi_bar2_reg;
-		break;
-	case PCI_BAR_ID_4:
-		bar_reg = pci.dbi_bar4_reg;
-		break;
-	default:
-		return -1;
-	}
-
-	aux_dbi_reg = regmap_readl(pci.aux_dbi_reg);
-	regmap_writel(IATU_ACCESS_DISABLE_VAL, pci.aux_dbi_reg);
-
-	bar_addr_low = regmap_readl(bar_reg) & ~0xF;
-	bar_addr_high = regmap_readl(bar_reg + sizeof(u32));
-
-	regmap_writel(aux_dbi_reg, pci.aux_dbi_reg);
-
-	*bar_phys_addr = (u64)bar_addr_high << 32 | bar_addr_low;
-
-	return 0;
-}
-
-/*
- * hl_pci_set_inbound_region() - Configure inbound region
- * @region: Inbound region number.
- * @pci_region: Inbound region parameters.
- *
- * Configure the iATU inbound region.
- */
-void hl_pci_set_inbound_region(u8 region,
-			       struct hl_pci_inbound_region *pci_region)
-{
-	u64 bar_phys_base = 0, region_base, region_end_address;
-	u32 offset = dev_pci_dbi_cfg_offset_get();
-	u32 ctrl_reg_val;
-	u32 aux_dbi_reg = regmap_readl(pci.aux_dbi_reg);
-
-	regmap_writel(IATU_ACCESS_ENABLE_VAL, pci.aux_dbi_reg);
-
-	/* region offset */
-	offset += (IATU_REGION_SIZE * region) + IATU_REGION_INBOUND_OFF;
-
-	/* disable region before configuring */
-	ctrl_reg_val = regmap_readl(offset + IATU_REGION_CTRL_2_OFF);
-	ctrl_reg_val &= ~IATU_REGION_EN_MASK;
-	regmap_writel(ctrl_reg_val, offset + IATU_REGION_CTRL_2_OFF);
-
-	if (pci_region->mode == PCI_ADDRESS_MATCH_MODE) {
-		hl_pci_get_bar_phys_addr(pci_region->bar, &bar_phys_base);
-		region_base = bar_phys_base + pci_region->offset_in_bar;
-		region_end_address = region_base + pci_region->size - 1;
-
-		regmap_writel((u32)region_base,
-			      offset + IATU_LWR_BASE_ADDR_OFF);
-		regmap_writel((u32)(region_base >> 32),
-			      offset + IATU_UPPER_BASE_ADDR_OFF);
-		regmap_writel((u32)region_end_address,
-			      offset + IATU_LIMIT_ADDR_OFF);
-	}
-
-	/* Point to the specified address */
-	regmap_writel((u32)pci_region->addr,
-		      offset + IATU_LWR_TARGET_ADDR_OFF);
-	regmap_writel((u32)(pci_region->addr >> 32),
-		      offset + IATU_UPPER_TARGET_ADDR_OFF);
-
-	/* Set bar type as memory */
-	regmap_writel(0, offset + IATU_REGION_CTRL_1_OFF);
-
-	/* Enable + bar/address match + match enable + bar number */
-	ctrl_reg_val = IATU_REGION_EN_MASK | IATU_NUM_MATCH_EN_MASK;
-
-	if (pci_region->mode == PCI_BAR_MATCH_MODE) {
-		ctrl_reg_val |= IATU_MATCH_MODE_MASK;
-		ctrl_reg_val |= (pci_region->bar << IATU_BAR_NUM_OFFSET) &
-				IATU_BAR_NUM_MASK;
-	}
-
-	regmap_writel(ctrl_reg_val, offset + IATU_REGION_CTRL_2_OFF);
-
-	/* restore iATU access state */
-	regmap_writel(aux_dbi_reg, pci.aux_dbi_reg);
-}
-
-/*
- * hl_pci_set_outbound_region() - Configure outbound region 0
- * @pci_region: Outbound region parameters.
- *
- * Configure the iATU outbound region 0.
- */
-void hl_pci_set_outbound_region(struct hl_pci_outbound_region *pci_region)
-{
-	u64 outbound_region_end_address;
-	u32 offset = dev_pci_dbi_cfg_offset_get();
-	u32 aux_dbi_reg = regmap_readl(pci.aux_dbi_reg);
-
-	regmap_writel(IATU_ACCESS_ENABLE_VAL, pci.aux_dbi_reg);
-
-	/* disable region before configuring */
-	regmap_writel(0, offset + IATU_REGION_CTRL_2_OFF);
-
-	/* Outbound Region 0 */
-	outbound_region_end_address =
-			pci_region->addr + pci_region->size - 1;
-	regmap_writel((u32)pci_region->addr, offset + IATU_LWR_BASE_ADDR_OFF);
-	regmap_writel((u32)(pci_region->addr >> 32),
-		      offset + IATU_UPPER_BASE_ADDR_OFF);
-	regmap_writel((u32)outbound_region_end_address,
-		      offset + IATU_LIMIT_ADDR_OFF);
-	regmap_writel(0, offset + IATU_LWR_TARGET_ADDR_OFF);
-	regmap_writel(0, offset + IATU_UPPER_TARGET_ADDR_OFF);
-
-	regmap_writel((u32)(outbound_region_end_address >> 32),
-		      offset + IATU_UPPR_LIMIT_ADDR_OFF);
-	/* Increase region size */
-	regmap_writel(IATU_OUTBOUND_REGION_SIZE,
-		      offset + IATU_REGION_CTRL_1_OFF);
-	/* Enable */
-	regmap_writel(IATU_REGION_EN_MASK, offset + IATU_REGION_CTRL_2_OFF);
-
-	/* restore iATU access state */
-	regmap_writel(aux_dbi_reg, pci.aux_dbi_reg);
-}
-
-/*
- * hl_pci_is_mem_enabled() - is memory access enabled in the configuration
- */
-__weak bool hl_pci_is_mem_enabled(void)
-{
-	u32 mem_space_ena_reg = regmap_readl(pci.aux_mem_space_en_reg);
-
-	return !!mem_space_ena_reg;
-}
-
-u32 hl_pci_get_ven_id_dev_id(void)
-{
-	u32 dev_id = preboot_device_id_get();
-
-	return (dev_id << 16) | PCI_VENDOR_ID_HABANALABS;
-}
-
-#define PCIE_DBI_STATUS_COMMAND_REG_PCI_TYPE0_MEM_SPACE_EN_SHIFT	1
-
-void hl_pcie_disable_mem_access(void)
-{
-	u32 reg;
-
-	/* Disable the PCIe inbound memory requests */
-	reg = regmap_readl(pci.dbi_status_command_reg);
-	reg &= ~BIT(PCIE_DBI_STATUS_COMMAND_REG_PCI_TYPE0_MEM_SPACE_EN_SHIFT);
-	regmap_writel(reg, pci.dbi_status_command_reg);
-}
-
-void hl_pcie_enable_mem_access(void)
-{
-	u32 reg;
-
-	/* Enable the PCIe inbound memory requests */
-	reg = regmap_readl(pci.dbi_status_command_reg);
-	reg |= BIT(PCIE_DBI_STATUS_COMMAND_REG_PCI_TYPE0_MEM_SPACE_EN_SHIFT);
-	regmap_writel(reg, pci.dbi_status_command_reg);
-}
-
-#define AER_UNCORR_ERR_MASK_CMPLT_ABORT_ERR_SHIFT 15
-
-void hl_pcie_mask_ca_aer(void)
-{
-	u32 uncorr_aer_mask_reg_val;
-
-	uncorr_aer_mask_reg_val = regmap_readl(pci.dbi_uncorr_aer_mask);
-	uncorr_aer_mask_reg_val |= BIT(AER_UNCORR_ERR_MASK_CMPLT_ABORT_ERR_SHIFT);
-	regmap_writel(uncorr_aer_mask_reg_val, pci.dbi_uncorr_aer_mask);
-}
-
-#define DBG_TIMEOUT_US		1000000
-#define DBG_LINE_SZ		80
-
-static unsigned int dbg_cnt;
-static u64 dbg_ts;
-
-__weak void pci_init_wait_dbg(void)
-{
-	if (timeout_expired(dbg_ts, DBG_TIMEOUT_US)) {
-		preboot_puts_raw_dbg("*");
-		dbg_ts = timestamp_get_us();
-		dbg_cnt++;
-	}
-
-	if (dbg_cnt == DBG_LINE_SZ) {
-		preboot_puts_dbg("\n");
-		dbg_cnt = 0;
-	}
-}
diff --git a/lib/common/pci.h b/lib/common/pci.h
deleted file mode 100644
index 91698e1..0000000
--- a/lib/common/pci.h
+++ /dev/null
@@ -1,31 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2017-2021 HabanaLabs Ltd.
- * All Rights Reserved.
- *
- */
-
-#ifndef _PCI_H_
-#define _PCI_H_
-
-#include <stdint.h>
-#include <stdbool.h>
-#include <types.h>
-#include <hl_pci_if.h>		/* requires stdint.h */
-
-void hl_pci_set_inbound_region(u8 region,
-			       struct hl_pci_inbound_region *pci_region);
-void hl_pci_set_outbound_region(struct hl_pci_outbound_region *pci_region);
-bool hl_pci_is_mem_enabled(void);
-u32 dev_pci_dbi_cfg_offset_get(void);
-u32 hl_pci_get_ven_id_dev_id(void);
-void hl_pcie_disable_mem_access(void);
-void hl_pcie_enable_mem_access(void);
-void hl_pcie_mask_ca_aer(void);
-void hl_pcie_init(void);
-void pci_init_wait_dbg(void);
-int hl_pci_set_bmu_region(struct hl_bmu_region *region);
-void hl_pci_bmu_set_cfg_base(u64 cfg_base_addr);
-void hl_pci_bmu_enable(bool enable);
-
-#endif /* _PCI_H_ */
diff --git a/lib/common/pll.c b/lib/common/pll.c
index 6c37b0b..e64e64e 100644
--- a/lib/common/pll.c
+++ b/lib/common/pll.c
@@ -85,6 +85,17 @@
 	return false;
 }
 
+/* Added these functions as wrapper for emb common modules usage */
+int64_t timer_get_us(void)
+{
+	return ((int64_t)timestamp_get_us());
+}
+
+bool timer_expired(int64_t start_tv, u32 delay)
+{
+	return timeout_expired((u64)start_tv, delay);
+}
+
 void udelay(unsigned int usec)
 {
 	u64 start, current, end = cycles_per_usec * usec;
diff --git a/lib/common/pll.h b/lib/common/pll.h
index 95d6407..ca32084 100644
--- a/lib/common/pll.h
+++ b/lib/common/pll.h
@@ -67,6 +67,8 @@
 u64 timestamp_get_us(void);
 bool timeout_expired(u64 start_tv, u32 delay);
 void hl_dev_pci_pll_wa(void);
+int64_t timer_get_us(void);
+bool timer_expired(int64_t start_tv, u32 delay);
 
 static inline void mdelay(unsigned int msec) { udelay(1000 * msec); }
 
diff --git a/lib/common/uart.h b/lib/common/uart.h
index 2919257..85a0b02 100644
--- a/lib/common/uart.h
+++ b/lib/common/uart.h
@@ -61,4 +61,11 @@
 		preboot_puts_hex(log, val, pad);
 }
 
+static inline void preboot_puts_hex64_dbg(const char *log,
+					unsigned long long val, bool pad)
+{
+	if (preboot_get_dbg_prints_ena())
+		preboot_puts_hex64(log, val, pad);
+}
+
 #endif /* _HL_UART_H_ */
diff --git a/lib/preboot/fit.c b/lib/preboot/fit.c
deleted file mode 100644
index 6613444..0000000
--- a/lib/preboot/fit.c
+++ /dev/null
@@ -1,1243 +0,0 @@
-/*
- * Copyright (C) 2017 HabanaLabs Ltd.
- *   Author: Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#include <libfdt.h>
-#include <string.h>
-#include <bitops.h>
-#include <security.h>
-#include <hl_boot_if.h>
-#include <baremetal.h>
-#include <hl_boot_desc.h> /* requires baremetal.h for __packed */
-#include <boot_if.h>
-#include <fit_load_type.h>
-#include <boot_info.h>
-#include <lzo/lzo.h>
-#include <hl_fw_if.h>
-#include <types.h>
-#include "uart.h"
-#include "fit.h"
-#include "fit_rsa.h"
-#include "fit_regions.h"
-#include "cmd_module.h"
-#include "i2cslave_indirect.h"
-#include <sha.h>
-
-#define FIT_IMAGES_PATH		"/images"
-#define FIT_CONFS_PATH		"/configurations"
-
-#define FIT_CONF_NAME		"recovery-conf"
-#define FIT_DESC_PROP		"description"
-#define FIT_LOADABLE_PROP	"loadables"
-#define FIT_HASHED_NODES_PROP	"hashed-nodes"
-#define FIT_DATA_PROP		"data"
-#define FIT_HASH_PROP		"hash"
-#define FIT_LOADADDR_PROP	"loadaddr"
-#define FIT_STARTADDR_PROP	"startaddr"
-#define FIT_ARCH_PROP		"arch"
-#define FIT_COMPRESSION_PROP	"compression"
-
-#define FIT_SEC_VERSION		"hl-sec-version"
-#define FIT_SEC_KEY_IDX		"hl-sec-key-idx"
-#define FIT_SIG_NODENAME	"signature"
-#define FIT_VALUE_PROP		"value"
-#define FIT_ALGO_PROP		"algo"
-#define FIT_KEY_NAME_HINT	"key-name-hint"
-#define FIT_PUB_KEY_PROP	"pub-key"
-#define FIT_PUB_KEY_EXT_PROP	"pub-key-ext"
-#define FIT_LOAD_TYPE_PROP	"load-type"
-#define FIT_PUB_KEY_HASH_PROP	"hash"
-
-#define FIT_MAX_PATH		200
-#define MAX_FDT_REGIONS		20
-#define IMAGE_MAX_HASHED_NODES	20
-
-#define FIT_MAX_IMAGES		8
-
-#define FDT_MIN_SUPPORTED_VERSION 0x11
-#define FIT_MAX_KEY_NAME	64
-#define PREFIX_KEY			"key-"
-
-#define FIT_ARC_PID_LOADABLE_PREFIX	"zephyr-pid"
-
-/*
- * For checking decompressed image bounderies, assume the compression factor to
- * be 5 times bigger than the compressed data.
- */
-#define COMPRESSION_FACTOR	5
-
-struct fit_image {
-	u64 loadaddr;
-	u64 startaddr;
-	const char *arch;
-	int core_id;
-};
-
-struct fit_image images[FIT_MAX_IMAGES];
-struct fdt_region fdt_regions[MAX_FDT_REGIONS];
-struct image_region region[MAX_FDT_REGIONS];
-
-__weak bool fit_is_core_id_supported(void)
-{
-	return false;
-}
-
-__weak void preboot_start_image(u64 startaddr, bool valid_arc_img)
-{
-	image_start(startaddr);
-}
-
-__weak void preboot_copy_image(u64 dst, const void *buf, int len,
-			       const char *arch)
-{
-	memcpy((void *)(uintptr_t)dst, buf, len);
-}
-
-__weak u64 preboot_get_arch_view_addr(u64 addr)
-{
-	return addr;
-}
-
-__weak int preboot_decompress_image(u64 dst, const void *buf, int len,
-				    const char *arch, const char *cmpr)
-{
-	void *dstptr = (void *)(uintptr_t)preboot_get_arch_view_addr(dst);
-	size_t dst_len = SIZE_MAX;
-	int err;
-
-	if (!strcmp(cmpr, "lzo")) {
-		preboot_puts_raw("LZO decompressing ");
-		err = lzop_decompress(buf, len, dstptr, &dst_len);
-		if (err)
-			preboot_puts_hex("error: ", -err, false);
-		else
-			preboot_puts_hex("done ", dst_len, false);
-
-		return err;
-	}
-
-	preboot_puts_raw("Unknown compression method: ");
-	preboot_puts_raw(cmpr);
-	preboot_puts("\n");
-
-	return 1;
-}
-
-__weak bool preboot_security_is_enabled(void)
-{
-	return false;
-}
-
-/*
- * Gaudi2 platform supports different bits in efuse for
- * security enable and image validation. To maintain
- * backward compatibility the default implementation
- * redirects to the security enable check.
- */
-__weak bool preboot_img_validation_is_enabled(void)
-{
-	return preboot_security_is_enabled();
-}
-
-__weak bool preboot_security_key_is_revoked(int key_idx)
-{
-	return false;
-}
-
-__weak bool preboot_security_key_is_valid(int key_idx, const void *pub_key,
-					  int pub_key_len,
-					  const void *pub_key_ext,
-					  int pub_key_ext_len)
-{
-	return false;
-}
-
-__weak bool preboot_security_svn_is_valid(int sec_ver)
-{
-	return false;
-}
-
-__weak bool preboot_security_key_montgomery_is_valid(void)
-{
-	return false;
-}
-
-static const char *fit_loadables_name(const void *fit, int noffset, int id)
-{
-	return fdt_stringlist_get(fit, noffset, FIT_LOADABLE_PROP, id, NULL);
-}
-
-static int fit_find_config_node(const void *fit)
-{
-	int conf, node;
-	int dflt_conf_node = -1;
-
-	conf = fdt_path_offset(fit, FIT_CONFS_PATH);
-	if (conf < 0) {
-		preboot_puts("Cannot find configuration node\n");
-		return -1;
-	}
-
-	for (node = fdt_first_subnode(fit, conf); node >= 0; node = fdt_next_subnode(fit, node)) {
-		const char *node_name = fdt_get_name(fit, node, NULL);
-
-		if (strcmp(FIT_CONF_NAME, node_name) == 0) {
-			dflt_conf_node = node;
-			break;
-		}
-	}
-
-	return dflt_conf_node;
-}
-
-static int fit_loadables_node_get(const void *fit, int id)
-{
-	int images, conf = fit_find_config_node(fit);
-	const char *uname;
-
-	if (conf == -1) {
-		preboot_puts("Wrong FIT type: Cannot find recovery conf\n");
-		return conf;
-	}
-
-	uname = fit_loadables_name(fit, conf, id);
-
-	images = fdt_path_offset(fit, FIT_IMAGES_PATH);
-	return fdt_subnode_offset(fit, images, uname);
-}
-
-int fit_check_format(const void *fit)
-{
-	if (fdt_check_header(fit))
-		return FIT_FORMAT_HEADER_CHECK_FAILED;
-
-	/* mandatory / node 'description' property */
-	if (fdt_getprop(fit, 0, FIT_DESC_PROP, NULL) == NULL)
-		return FIT_FORMAT_NO_DESCRIPTION;
-
-	/* mandatory subimages parent '/images' node */
-	if (fdt_path_offset(fit, FIT_IMAGES_PATH) < 0)
-		return FIT_FORMAT_NO_IMAGES_PARENT_NODE;
-
-	return FIT_FORMAT_SUCCESS;
-}
-
-static int fit_get_int(const void *fit, int node, const char *prop_name,
-		       int default_val)
-{
-	const int *cell;
-	int len;
-
-	cell = fdt_getprop(fit, node, prop_name, &len);
-	if (cell && len >= sizeof(int))
-		return fdt32_to_cpu(cell[0]);
-
-	return default_val;
-}
-
-static int get_algo_prop(const void *fit, int noffset, const u8 **algo_prop)
-{
-	int algo_prop_len;
-
-	*algo_prop = fdt_getprop(fit, noffset, FIT_ALGO_PROP, &algo_prop_len);
-	if (!(*algo_prop)) {
-		preboot_puts("Wrong FIT format: algo property\n");
-		return 1;
-	}
-	if (algo_prop_len != sizeof(FIT_ALGO_PROP_EXPECTED_VALUE)) {
-		preboot_puts("Wrong FIT format: algo len\n");
-		return 1;
-	}
-	if (memcmp((char *)*algo_prop, FIT_ALGO_PROP_EXPECTED_VALUE,
-			sizeof(FIT_ALGO_PROP_EXPECTED_VALUE))) {
-		preboot_puts("Wrong FIT format: hash-algo not supported\n");
-		preboot_puts((char *)algo_prop);
-		preboot_puts("\n");
-		return 1;
-	}
-	return 0;
-}
-
-static int get_sig_hash(const void *fit, int noffset, const u8 **hash_value,
-			int *hash_value_len)
-{
-	*hash_value = fdt_getprop(fit, noffset, FIT_VALUE_PROP, hash_value_len);
-	if (!(*hash_value)) {
-		preboot_puts("Wrong FIT format: missing sig hash value\n");
-		return 1;
-	}
-	return 0;
-}
-
-static int get_key_name(const void *fit, int noffset, const char **key_name)
-{
-	int key_name_len;
-
-	*key_name = fdt_getprop(fit, noffset, FIT_KEY_NAME_HINT, &key_name_len);
-	if (!(*key_name)) {
-		preboot_puts("Wrong FIT format: key name\n");
-		return 1;
-	}
-	return 0;
-}
-
-static int get_pub_key_node(const void *pub_key_fit, const char *key_name)
-{
-	int sig_node;
-	char key_node_name[FIT_MAX_KEY_NAME] = PREFIX_KEY;
-	size_t key_node_name_len = sizeof(PREFIX_KEY) - 1;
-
-	key_node_name[FIT_MAX_KEY_NAME - 1] = '\0';
-
-	strncpy(key_node_name + key_node_name_len,
-			key_name, FIT_MAX_KEY_NAME - key_node_name_len);
-	/*
-	 * To avoid stack based buffer overflow
-	 * make sure your key string length does not exceed buffer size.
-	 */
-	if (key_node_name[FIT_MAX_KEY_NAME - 1] != '\0')
-		return -1;
-
-	sig_node = fdt_subnode_offset(pub_key_fit, 0, FIT_SIG_NODENAME);
-
-	return fdt_subnode_offset(pub_key_fit, sig_node, key_node_name);
-}
-
-int fit_get_key_prop(const void *pub_key_fit, const char *key_name,
-			    struct rsa_key_prop *prop)
-{
-	int pub_key_node = 0;
-	int length;
-
-	if (fdt_check_header(pub_key_fit)) {
-		preboot_puts("FITkey header check failed\n");
-		return 1;
-	}
-	pub_key_node = get_pub_key_node(pub_key_fit, key_name);
-	if (pub_key_node < 0) {
-		preboot_puts("Wrong FITkey format: no key node\n");
-		return 1;
-	}
-	prop->num_bits =
-		fit_get_int(pub_key_fit, pub_key_node, "rsa,num-bits", 0);
-	prop->n0inv =
-		fit_get_int(pub_key_fit, pub_key_node, "rsa,n0-inverse", 0);
-	prop->public_exponent =
-		fdt_getprop(pub_key_fit, pub_key_node, "rsa,exponent", &length);
-	if (!prop->public_exponent || length < sizeof(uint64_t))
-		prop->public_exponent = NULL;
-	prop->modulus =
-		fdt_getprop(pub_key_fit, pub_key_node, "rsa,modulus", NULL);
-	prop->rr =
-		fdt_getprop(pub_key_fit, pub_key_node, "rsa,r-squared", NULL);
-
-	if (!prop->num_bits || !prop->modulus) {
-		preboot_puts("FIT: missing RSA Key\n");
-		return 1;
-	}
-
-	return 0;
-}
-
-static int fit_config_check_sig(const void *fit, int noffset,
-				const struct image_region region[],
-				int region_count, const void *pub_key_fit)
-{
-	const u8 *algo_prop = NULL;
-	const u8 *hash_value = NULL;
-	const char *key_name = NULL;
-	int hash_len = 0;
-	struct rsa_key_prop prop;
-
-	if (get_algo_prop(fit, noffset, &algo_prop))
-		return 1;
-
-	if (get_sig_hash(fit, noffset, &hash_value, &hash_len))
-		return 1;
-
-	if (get_key_name(fit, noffset, &key_name))
-		return 1;
-
-	if (fit_get_key_prop(pub_key_fit, key_name, &prop))
-		return 1;
-
-	return rsa_verify_regions(region, region_count, hash_value, hash_len,
-				  &prop);
-}
-
-static const char *fit_get_hashed_nodes(const void *fit, int noffset,
-					int *hashed_node_count, char **node_inc)
-{
-	const char *prop, *end, *name;
-	int prop_len;
-	int count;
-	int i;
-
-	/* Count the number of strings in the property */
-	count = fdt_stringlist_count(fit, noffset, FIT_HASHED_NODES_PROP);
-	if ((count == 0) || (count > IMAGE_MAX_HASHED_NODES)) {
-		preboot_puts_hex("FIT: bad hashed-node: ", count, false);
-		return NULL;
-	}
-
-	prop = fdt_getprop(fit, noffset, FIT_HASHED_NODES_PROP, &prop_len);
-	end = prop ? prop + prop_len : prop;
-	*hashed_node_count = count;
-	for (name = prop, i = 0; name < end; name += strlen(name) + 1, i++)
-		node_inc[i] = (char *)name;
-
-	return prop;
-}
-
-static int fit_pub_key_node_get(const void *fit, int conf)
-{
-	int images;
-
-	images = fdt_path_offset(fit, FIT_IMAGES_PATH);
-	return fdt_subnode_offset(fit, images, FIT_PUB_KEY_PROP);
-}
-
-static int fit_pub_key_ext_node_get(const void *fit, int conf)
-{
-	int images;
-
-	images = fdt_path_offset(fit, FIT_IMAGES_PATH);
-	return fdt_subnode_offset(fit, images, FIT_PUB_KEY_EXT_PROP);
-}
-
-__weak void preboot_security_boot_info_write_key_idx(u32 *key_idx) {}
-
-static u32 preboot_key_idx;
-
-static int fit_get_pub_key(const void *fit, int noffset, int sig_sub_node,
-			   const void **pub_key_fit, enum fit_sec_failure *res)
-{
-	int node, key_len, pub_key_node;
-	const void *key_ext_data = NULL;
-	const void *key_data;
-	const char *key_name;
-	int key_ext_len = 0;
-
-	node = fit_pub_key_node_get(fit, noffset);
-	if (node < 0) {
-		preboot_puts("Wrong FIT format missing pubkey\n");
-		return 1;
-	}
-
-	key_data = fdt_getprop(fit, node, FIT_DATA_PROP, &key_len);
-	if (!key_data) {
-		preboot_puts("Wrong FIT format: missing pubkey fit\n");
-		return 1;
-	}
-
-	preboot_key_idx = fit_get_int(fit, 0, FIT_SEC_KEY_IDX, -1);
-	if (preboot_key_idx < 0) {
-		preboot_puts("Wrong FIT format: missing key idx\n");
-		return 1;
-	}
-
-	*pub_key_fit = key_data;
-	if (!preboot_security_key_montgomery_is_valid())
-		goto key_validation;
-
-	/* Get modulus from public key */
-	if (fdt_check_header(*pub_key_fit)) {
-		preboot_puts("FITkey header check failed\n");
-		return 1;
-	}
-
-	if (get_key_name(fit, sig_sub_node, &key_name))
-		return 1;
-
-	pub_key_node = get_pub_key_node(*pub_key_fit, key_name);
-	if (pub_key_node < 0) {
-		preboot_puts("Wrong FITkey format: no key node\n");
-		return 1;
-	}
-
-	/* Update key data and key length with 384 bytes modulus */
-	key_data = fdt_getprop(*pub_key_fit, pub_key_node,
-			"rsa,modulus", &key_len);
-
-	/* Get montgomery from public key extension */
-	node = fit_pub_key_ext_node_get(fit, noffset);
-	if (node < 0) {
-		preboot_puts("Wrong FIT format missing ext pubkey\n");
-		return 1;
-	}
-
-	key_ext_data = fdt_getprop(fit, node, FIT_DATA_PROP,
-			&key_ext_len);
-	if (!key_ext_data) {
-		preboot_puts("Wrong FIT format: missing ext pubkey fit\n");
-		return 1;
-	}
-
-key_validation:
-	if (preboot_security_key_is_revoked(preboot_key_idx)) {
-		preboot_puts_hex("Key revoked: ", preboot_key_idx, false);
-		*res = FIT_SEC_KEY_REVOKED;
-		return 1;
-	}
-
-	if (!preboot_security_key_is_valid(preboot_key_idx, key_data, key_len,
-				key_ext_data, key_ext_len)) {
-		preboot_puts("Key validation failed\n");
-		*res = FIT_SEC_SIGN_FAILED;
-		return 1;
-	}
-
-	return 0;
-}
-
-static bool fit_svn_is_valid(const void *fit)
-{
-	int sec_ver;
-
-	sec_ver = fit_get_int(fit, 0, FIT_SEC_VERSION, -1);
-	if (sec_ver < 0) {
-		preboot_puts("Wrong FIT format: missing security version\n");
-		return false;
-	}
-
-	return preboot_security_svn_is_valid(sec_ver);
-}
-
-static int fit_verify_config_regions(const void *fit,
-				     struct fdt_region *fdt_regions,
-				     int count, int sig_sub_node,
-				     const void *pub_key_fit)
-{
-	fit_region_make_list(fit, fdt_regions, count, region);
-
-	return fit_config_check_sig(fit, sig_sub_node, region, count,
-				    pub_key_fit);
-}
-
-/* Checks the signed FIT and validate it againts the public key */
-static int fit_verify_config(const void *fit, int noffset,
-			     enum fit_sec_failure *res)
-{
-	int sig_sub_node;
-	const char *hashed_node_prop;
-	int count, err;
-	char *node_inc[IMAGE_MAX_HASHED_NODES];
-	char * const exc_prop[] = { FIT_DATA_PROP };
-	char path[FIT_MAX_PATH];
-	const uint32_t *strings;
-	const void *pub_key_fit;
-
-	sig_sub_node = fdt_subnode_offset(fit, noffset, FIT_SIG_NODENAME);
-	if (sig_sub_node < 0) {
-		preboot_puts("Wrong FIT format: missing signature\n");
-		return 1;
-	}
-
-	if (fit_get_pub_key(fit, noffset, sig_sub_node, &pub_key_fit, res))
-		return 1;
-
-	hashed_node_prop = fit_get_hashed_nodes(fit, sig_sub_node, &count,
-						node_inc);
-	if (!hashed_node_prop)
-		return 1;
-
-	/* Get a list of regions to hash */
-	count = fdt_find_regions(fit, node_inc, count, exc_prop,
-				 ARRAY_SIZE(exc_prop), fdt_regions,
-				 MAX_FDT_REGIONS - 1, path, sizeof(path), 0);
-
-	if ((count <= 0) || (count >= MAX_FDT_REGIONS - 1)) {
-		preboot_puts_hex("FIT: bad regions count: ", count, 0);
-		return 1;
-	}
-
-	/* Add the strings */
-	strings = fdt_getprop(fit, sig_sub_node, "hashed-strings", NULL);
-	if (strings) {
-		fdt_regions[count].offset = fdt_off_dt_strings(fit);
-		fdt_regions[count].size = fdt32_to_cpu(strings[1]);
-		count++;
-	}
-
-	err = fit_verify_config_regions(fit, fdt_regions, count, sig_sub_node,
-					pub_key_fit);
-	if (err) {
-		*res = FIT_SEC_SIGN_FAILED;
-		return err;
-	}
-
-	return 0;
-}
-
-static int fit_verify_security(const void *fit, int noffset,
-			       enum fit_sec_failure *res)
-{
-	if (!preboot_img_validation_is_enabled())
-		return 0;
-
-	if (!fit_svn_is_valid(fit)) {
-		*res = FIT_SEC_SVN_INVALID;
-		return 1;
-	}
-
-	/* Security in enabled and SV is valid all thats left is to check the
-	 * signature on the FIT.
-	 */
-	return fit_verify_config(fit, noffset, res);
-}
-
-static int fit_getprop_u64(const void *fit, int node, const char *name,
-			   u64 *val64)
-{
-	const u32 *val32;
-	int len;
-
-	val32 = fdt_getprop(fit, node, name, &len);
-	if (!val32 || len != sizeof(u64))
-		return 1;
-
-	*val64 = ((u64)fdt32_to_cpu(val32[0])) << 32;
-	*val64 |= fdt32_to_cpu(val32[1]);
-	return 0;
-}
-
-int fit_init_resp(void *fit, size_t sz)
-{
-	int err;
-
-	err = fdt_create(fit, sz);
-	if (err)
-		return err;
-
-	err = fdt_finish_reservemap(fit);
-	if (err)
-		return err;
-
-	err = fdt_begin_node(fit, "");
-	if (err)
-		return err;
-
-	return 0;
-}
-
-int fit_fin_resp(void *fit, unsigned int *res_size)
-{
-	int err;
-
-	if (!fdt_size_dt_strings(fit)) {
-		*res_size = 0;
-		return 0; /* Empty response */
-	}
-
-	err = fdt_end_node(fit);
-	if (err)
-		return err;
-
-	err = fdt_finish(fit);
-	if (err)
-		return err;
-
-	*res_size = fdt_totalsize(fit);
-
-	return err;
-}
-
-__weak int fit_process_core_id(enum fit_core_id core_id)
-{
-	return 1;
-}
-
-static int fit_copy_image(u64 dst, const void *buf, int len, const char *arch,
-			  const char *cmpr)
-{
-	preboot_puts_hex64("FIT loadaddr: ", dst, false);
-
-	if (!strcmp(cmpr, "none")) {
-		preboot_copy_image(dst, buf, len, arch);
-		return 0;
-	}
-
-	return preboot_decompress_image(dst, buf, len, arch, cmpr);
-}
-
-static bool fit_load_is_runnable(const void *fit, int conf)
-{
-	const u32 *cell;
-	u32 type;
-	int len;
-
-	cell = fdt_getprop(fit, conf, FIT_LOAD_TYPE_PROP, &len);
-
-	/* If load type does not exists, assume it's u-boot */
-	if (!cell)
-		return true;
-
-	if (len != sizeof(*cell)) {
-		preboot_puts("Corrupted load type\n");
-		return false;
-	}
-
-	type = fdt32_to_cpu(*cell);
-	if (type >= FIT_LOAD_TYPE_COUNT) {
-		preboot_puts_hex("Unsupported load type ", type, false);
-		return false;
-	}
-
-	return type == FIT_LOAD_TYPE_RUN;
-}
-
-static const u8 *fit_node_get_hash(const void *fit, int node)
-{
-	int hash;
-	const char *prop;
-
-	hash = fdt_subnode_offset(fit, node, FIT_HASH_PROP);
-	prop = (const char *)fdt_getprop(fit, hash, FIT_ALGO_PROP, NULL);
-	if (strcmp(prop, FIT_SHA))
-		return NULL;
-
-	return (const u8 *)fdt_getprop(fit, hash, FIT_VALUE_PROP, NULL);
-}
-
-__weak void sha_csum(const unsigned char *input, unsigned int ilen,
-		unsigned char *output)
-{
-}
-
-static int fit_node_check_hash(const void *fit, int node, const u8 *buf,
-			       int len)
-{
-	const u8 *node_hash = fit_node_get_hash(fit, node);
-	u8 calc_hash[SHA_SUM_LEN];
-
-	if (!node_hash) {
-		preboot_puts("Missing hash for loadable\n");
-		return 1;
-	}
-
-	sha_csum(buf, len, calc_hash);
-	if (memcmp(node_hash, calc_hash, SHA_SUM_LEN))
-		return 1;
-
-	return 0;
-}
-
-__weak bool preboot_is_image_sig_supported(void)
-{
-	return false;
-}
-
-__weak int preboot_verify_img_sig(const void *fit, size_t img_size,
-				  enum fit_sec_failure *res, u32 *key_idx)
-{
-	return 1;
-}
-
-static int fit_verify_image(const void *fit, size_t fit_size, int conf,
-			    enum fit_sec_failure *res)
-{
-	/*
-	 * for platforms with image security descriptor support, the signature
-	 * of the descriptor is validated. otherwise the fit internal signature
-	 * is validated.
-	 */
-	if (!preboot_is_image_sig_supported())
-		return fit_verify_security(fit, conf, res);
-
-	return preboot_verify_img_sig(fit, fit_size, res, &preboot_key_idx);
-}
-
-/*
- * For secureboot flow, ARC0 is used to run
- * PCIe initialization code and put in sleep
- * once initialization is done. ARC0 will be
- * halt and reseted as part of secondary_cpu_stop
- * function before copying images in ICCM and DCCM
- * as part of fit_copy_image function.
- */
-__weak void preboot_secondary_cpu_stop(void) {}
-
-static bool fit_image_is_for_arc0(const void *fit, int id)
-{
-	const char *loadable_name;
-
-	loadable_name = fit_loadables_name(fit, fit_find_config_node(fit), id);
-	if (!loadable_name) {
-		preboot_puts_hex("Wrong FIT format: loadable name not found: ", id, false);
-		return false;
-	}
-
-	/* Note: We assume that the image that is intended to run on ARC0 is ARC PID */
-	return !(memcmp(loadable_name, FIT_ARC_PID_LOADABLE_PREFIX,
-			sizeof(FIT_ARC_PID_LOADABLE_PREFIX) - 1));
-}
-
-static enum fit_load_result fit_copy_images(const void *fit,
-					    enum fit_sec_failure *res)
-{
-	int node, len, image_len, core_id_len;
-	int conf, num_images, i;
-	const u8 *buf;
-	u64 load, start;
-	const char *cmpr;
-	size_t fit_size = fdt_totalsize(fit);
-	bool halted_secondary_cpu = false;
-	const uint32_t *core_id;
-
-	/* Check if image supports a minimum FDT version or newer */
-	if (fdt_version(fit) < FDT_MIN_SUPPORTED_VERSION)
-		return FIT_LOAD_SEC_FAILED;
-
-	/* check that the calculated memory range are legal */
-	if (!preboot_sram_is_valid_address_range((uintptr_t)fit, fit_size))
-		return FIT_LOAD_SEC_FAILED;
-
-	/* Get number of images in preboot configuration */
-	conf = fit_find_config_node(fit);
-	if (conf < 0)
-		return FIT_LOAD_CORRUPTED;
-
-	if (!fit_load_is_runnable(fit, conf))
-		return FIT_LOAD_CORRUPTED;
-
-	if (fit_verify_image(fit, fit_size, conf, res))
-		return FIT_LOAD_SEC_FAILED;
-
-	num_images = fdt_stringlist_count(fit, conf, FIT_LOADABLE_PROP);
-	if (num_images > FIT_MAX_IMAGES) {
-		preboot_puts_hex("Images in fit exceeds max limit: ",
-					FIT_MAX_IMAGES, false);
-		return FIT_LOAD_CORRUPTED;
-	} else if (num_images < 0) {
-		preboot_puts_hex("Cant read images from fit: ",
-					-num_images, false);
-		return FIT_LOAD_CORRUPTED;
-	}
-
-	for (i = 0; i < num_images; i++) {
-		node = fit_loadables_node_get(fit, i);
-		if (node < 0) {
-			preboot_puts_hex("Wrong FIT format: loadable image not found: ", i, false);
-			return FIT_LOAD_NOT_RUNNABLE;
-		}
-
-		if (fit_getprop_u64(fit, node, FIT_LOADADDR_PROP, &load)) {
-			preboot_puts_hex("Wrong FIT format: loadaddr not found for image: ", i, false);
-			return FIT_LOAD_BAD_LOAD_ADDR;
-		}
-		if (fit_getprop_u64(fit, node, FIT_STARTADDR_PROP, &start))
-			start = load;
-
-		buf = fdt_getprop(fit, node, FIT_DATA_PROP, &len);
-		if (!buf) {
-			preboot_puts_hex("Wrong FIT format: data empty for image: ", i, false);
-			return FIT_LOAD_CORRUPTED;
-		}
-
-		if (fit_node_check_hash(fit, node, buf, len)) {
-			preboot_puts_hex("FIT loadable corrupted: bad hash for image: ", i, false);
-			return FIT_LOAD_CORRUPTED;
-		}
-
-		cmpr = (const char *)fdt_getprop(fit, node,
-						 FIT_COMPRESSION_PROP,
-						 NULL);
-		if (!cmpr) {
-			/*
-			 * mkimage is adding compression, if it's missing then
-			 * something is wrong.
-			 */
-			preboot_puts_hex("Wrong FIT format: missing compression: ", i, false);
-			return FIT_LOAD_CORRUPTED;
-		}
-
-		if (!strcmp(cmpr, "none"))
-			image_len = len;
-		else
-			image_len = len * COMPRESSION_FACTOR;
-
-		if (!preboot_is_valid_address_range(load, image_len)) {
-			preboot_puts_hex("FIT: invalid load address for image: ", i, false);
-			return FIT_LOAD_BAD_LOAD_ADDR;
-		}
-
-		images[i].arch = (const char *)fdt_getprop(fit, node,
-							FIT_ARCH_PROP, NULL);
-		if (!images[i].arch) {
-			preboot_puts_hex("Wrong FIT format: arch empty for image: ", i, false);
-			return FIT_LOAD_CORRUPTED;
-		}
-
-		if (!halted_secondary_cpu && fit_image_is_for_arc0(fit, i)) {
-			if (!is_pcie_initialized())
-				preboot_puts("Warning: ARC0 busy with PCI init, halting anyway\n");
-
-			preboot_secondary_cpu_stop();
-			halted_secondary_cpu = true;
-		}
-
-		images[i].loadaddr = load;
-		images[i].startaddr = start;
-
-		if (fit_is_core_id_supported()) {
-			core_id = fdt_getprop(fit, node, HL_FIT_CORE_ID_PROP, &core_id_len);
-			if (!core_id)
-				return FIT_LOAD_CORRUPTED;
-
-			if (core_id_len != sizeof(*core_id))
-				return FIT_LOAD_CORRUPTED;
-
-			images[i].core_id = fdt32_to_cpu(*core_id);
-			if (fit_process_core_id((enum fit_core_id)images[i].core_id))
-				return FIT_LOAD_CORRUPTED;
-		}
-
-		if (fit_copy_image(load, buf, len, images[i].arch, cmpr))
-			return FIT_LOAD_CORRUPTED;
-	}
-
-	preboot_security_boot_info_write_key_idx(&preboot_key_idx);
-
-	return FIT_LOAD_SUCCESS;
-}
-
-static void fit_report_err(enum fit_load_result err,
-			   enum fit_sec_failure sec_res)
-{
-	enum hl_fw_fit_boot_error boot_err = HL_FW_FIT_BOOT_ERR_NONE;
-
-	switch (err) {
-	case FIT_LOAD_BAD_LOAD_ADDR:
-		boot_err = HL_FW_FIT_BOOT_ERR_ADDRESS_VIOLATION;
-		break;
-	case FIT_LOAD_NOT_RUNNABLE:
-		boot_err = HL_FW_FIT_BOOT_ERR_NOT_RUNNABLE;
-		break;
-	case FIT_LOAD_CORRUPTED:
-		boot_err = HL_FW_FIT_BOOT_ERR_CORRUPTED;
-		break;
-	case FIT_LOAD_SEC_FAILED:
-
-		switch (sec_res) {
-		case FIT_SEC_SVN_INVALID:
-			boot_err = HL_FW_FIT_BOOT_ERR_SVN;
-			break;
-		case FIT_SEC_KEY_REVOKED:
-			boot_err = HL_FW_FIT_BOOT_ERR_REVOKED_KEY;
-			break;
-		case FIT_SEC_SIGN_FAILED:
-			boot_err = HL_FW_FIT_BOOT_ERR_SIGNATURE;
-			break;
-		default:
-			break;
-		}
-
-		break;
-	default:
-		break;
-	};
-
-	oob_set_boot_err(boot_err, false);
-}
-
-static int fwupd_sec_convert_min_error(enum fit_sec_failure err)
-{
-	int serr;
-
-	switch (err) {
-	case FIT_SEC_KEY_REVOKED:
-		serr = HL_SEC_ERR_KEY_REVOKED;
-		break;
-	case FIT_SEC_SVN_INVALID:
-		serr = HL_SEC_ERR_SVN_INVALID;
-		break;
-	case FIT_SEC_SIGN_FAILED:
-		serr = HL_SEC_ERR_SIGN_FAILED;
-		break;
-	default:
-		serr = FIT_SEC_NONE;
-	};
-
-	return serr;
-}
-
-void fit_load_print_error(enum fit_load_result errtype)
-{
-	switch (errtype) {
-	case FIT_LOAD_BAD_LOAD_ADDR:
-	case FIT_LOAD_NOT_RUNNABLE:
-	case FIT_LOAD_CORRUPTED:
-		preboot_puts("FIT loading failed\n");
-		break;
-	case FIT_LOAD_SEC_FAILED:
-		preboot_puts("FIT security validation failed\n");
-		break;
-	default:
-		preboot_puts_hex("unrecognized error value for: fit_load_result",
-			(unsigned int)errtype, false);
-		break;
-	}
-}
-
-void fit_format_print_error(enum fit_format_result errtype)
-{
-	switch (errtype) {
-	case FIT_FORMAT_HEADER_CHECK_FAILED:
-		preboot_puts("FIT header check failed\n");
-		break;
-	case FIT_FORMAT_NO_DESCRIPTION:
-		preboot_puts("Wrong FIT format: no description\n");
-		break;
-	case FIT_FORMAT_NO_IMAGES_PARENT_NODE:
-		preboot_puts("Wrong FIT format: no images parent node\n");
-		break;
-	case FIT_FORMAT_CURRUPTED_LOAD_TYPE:
-		preboot_puts("Currupted load type\n");
-		break;
-	case FIT_FORMAT_UNSUPPORTED_LOAD_TYPE:
-		preboot_puts("Unsupported load type\n");
-		break;
-	default:
-		preboot_puts_hex("unrecognized error value for: fit_format_result",
-			(unsigned int)errtype, false);
-		break;
-	}
-}
-
-int fit_load_images(void *fit)
-{
-	enum fit_sec_failure sec_res = FIT_SEC_NONE;
-	enum fit_load_result fit_load_error;
-	enum fit_format_result fit_format_error;
-
-	memset(&images, 0x0, sizeof(images));
-
-	fit_format_error = fit_check_format(fit);
-	if (fit_format_error) {
-		fit_format_print_error(fit_format_error);
-		fit_load_error = FIT_LOAD_CORRUPTED;
-	} else {
-		fit_load_error = fit_copy_images(fit, &sec_res);
-	}
-
-
-	cpu_boot_err_clrset(BIT_ULL(CPU_BOOT_ERR_FIT_CORRUPTED) |
-			    BIT_ULL(CPU_BOOT_ERR_SECURITY_FAIL), false);
-	switch (fit_load_error) {
-	case FIT_LOAD_SUCCESS:
-		return FIT_LOAD_SUCCESS;
-	case FIT_LOAD_SEC_FAILED:
-		cpu_boot_err_set(CPU_BOOT_ERR_SECURITY_FAIL);
-		preboot_error_inform(HL_FW_ERR_SECURITY,
-				     fwupd_sec_convert_min_error(sec_res));
-		break;
-	default:
-		cpu_boot_err_set(CPU_BOOT_ERR_FIT_CORRUPTED);
-		preboot_error_inform(HL_FW_ERR_FIT, fit_load_error);
-		break;
-	}
-
-	fit_report_err(fit_load_error, sec_res);
-	fit_load_print_error(fit_load_error);
-
-	return fit_load_error;
-}
-
-__weak void preboot_finalize_fw_info(void) {}
-__weak void preboot_finalize_rma_tables(void) {}
-__weak void preboot_finalize_fw_upd_closure(void) {}
-
-void fit_run_image(void)
-{
-	u64 startaddr = 0;
-	bool valid_arc_img = false;
-	u32 i;
-
-	for (i = 0; i < FIT_MAX_IMAGES; i++) {
-
-		/*
-		 * Below check to detect end of valid images
-		 * in images array.
-		 */
-		if (!images[i].arch)
-			break;
-
-		if (!(strcmp(images[i].arch, "arm64")))
-			startaddr = images[i].startaddr;
-
-		if (!(strcmp(images[i].arch, "arc"))) {
-			valid_arc_img = true;
-
-			if (fit_is_core_id_supported()) {
-				/* start ARC1 image in end, so skip running arc1 image here */
-				if (!(fit_process_core_id(images[i].core_id))) {
-					startaddr = images[i].startaddr;
-					continue;
-				}
-			}
-		}
-
-		if ((valid_arc_img) && (startaddr))
-			break;
-	}
-
-	preboot_finalize_nic_en();
-	preboot_finalize_fw_info();
-	preboot_finalize_rma_tables();
-	preboot_finalize_fw_upd_closure();
-	preboot_finalize_enabled_caps();
-	preboot_finalize_uboot_override();
-	hl_boot_info_finalize();
-	preboot_start_image(startaddr, valid_arc_img);
-}
-
-__weak size_t preboot_get_img_sec_desc_size(void)
-{
-	return 0;
-}
-
-int fit_validate_size(const void *fit, int len)
-{
-	size_t img_size = ROUND_UP4(fdt_totalsize(fit));
-
-	if (fdt_check_header(fit)) {
-		preboot_puts("Bad FIT header\n");
-		return 1;
-	}
-
-	/* taking the security footer into account */
-	img_size += preboot_get_img_sec_desc_size();
-
-	return img_size != len;
-}
-
-int fit_get_load_type(const void *fit, u32 *load_type)
-{
-	const u32 *cell;
-	u32 type = FIT_LOAD_TYPE_RUN;
-	int ret, len, conf;
-
-	ret = fit_check_format(fit);
-	if (ret)
-		return ret;
-
-	conf = fit_find_config_node(fit);
-	if (conf >= 0) {
-		cell = fdt_getprop(fit, conf, FIT_LOAD_TYPE_PROP, &len);
-		if (!cell) {
-			preboot_puts("No load type. Assuming old FIT version\n");
-			*load_type = type;
-			return FIT_FORMAT_SUCCESS;
-		}
-		if (len != sizeof(*cell))
-			return FIT_FORMAT_CURRUPTED_LOAD_TYPE;
-
-		type = fdt32_to_cpu(*cell);
-		if (type >= FIT_LOAD_TYPE_COUNT)
-			return FIT_FORMAT_UNSUPPORTED_LOAD_TYPE;
-	}
-
-	*load_type = type;
-
-	return ret;
-}
-
-static int fit_image_get_data(const void *fit, int noffset,
-		       const void **data, size_t *size)
-{
-	int len;
-
-	*data = fdt_getprop(fit, noffset, FIT_DATA_PROP, &len);
-	if (*data == NULL) {
-		*size = 0;
-		return -1;
-	}
-
-	*size = len;
-	return 0;
-}
-
-static int fit_image_get_node(const void *fit, const char *image_uname)
-{
-	int noffset, images_noffset;
-
-	images_noffset = fdt_path_offset(fit, FIT_IMAGES_PATH);
-	if (images_noffset < 0) {
-		preboot_puts("Can't find images parent node\n");
-		return images_noffset;
-	}
-
-	noffset = fdt_subnode_offset(fit, images_noffset, image_uname);
-	if (noffset < 0)
-		preboot_puts("Can't get node offset for image unit name\n");
-
-	return noffset;
-}
-
-static int fit_conf_get_prop_node_index(const void *fit, int noffset,
-				 const char *prop_name, int index)
-{
-	const char *uname;
-	int len;
-
-	/* get image unit name from configuration property */
-	uname = fdt_stringlist_get(fit, noffset, prop_name, index, &len);
-	if (uname == NULL)
-		return len;
-
-	return fit_image_get_node(fit, uname);
-}
-
-int fit_image_get_loadable_data(const void *fit, const void **data,
-				size_t size)
-{
-	size_t actual_size;
-	int err, node, conf;
-
-	conf = fit_find_config_node(fit);
-	if (conf < 0)
-		return 1;
-
-	node = fit_conf_get_prop_node_index(fit, conf, FIT_LOADABLE_PROP, 0);
-	if (node < 0)
-		return 1;
-
-	err = fit_image_get_data(fit, node, data, &actual_size);
-	if (err)
-		return err;
-
-	if (actual_size != size)
-		return 1;
-
-	return 0;
-}
-
-#define HL_SEC_REGIONS_CNT	2
-#define HL_SEC_KEY_NAME		"hl_rsa3072"
-
-int fit_verify_img_sig(const void *img, size_t img_size, const void *sec_desc,
-		       size_t sec_desc_size, const void *pub_key_fit,
-		       const u8 *sig, size_t sig_len)
-{
-	const char *key_name = HL_SEC_KEY_NAME;
-	struct rsa_key_prop prop;
-	struct image_region region[HL_SEC_REGIONS_CNT];
-
-	if (fit_get_key_prop(pub_key_fit, key_name, &prop))
-		return 1;
-
-	/*
-	 * region 0 holds the fit image as a whole,
-	 * region 1 holds the hl security descriptor
-	 */
-	region[0].data = img;
-	region[0].size = img_size;
-	region[1].data = sec_desc;
-	region[1].size = sec_desc_size;
-
-	return rsa_verify_regions(region, HL_SEC_REGIONS_CNT, sig, sig_len,
-				  &prop);
-}
diff --git a/lib/preboot/fit.h b/lib/preboot/fit.h
deleted file mode 100644
index d35cc11..0000000
--- a/lib/preboot/fit.h
+++ /dev/null
@@ -1,48 +0,0 @@
-/*
- * Copyright (C) 2017 HabanaLabs Ltd.
- *   Author: Noam Camus <noamikong@gmail.com>
- *
- * SPDX-License-Identifier:     GPL-2.0+
- */
-
-#ifndef _HL_FIT_H_
-#define _HL_FIT_H_
-
-#include <hl_fit_if.h>
-
-#define HL_SEC_KEY_NAME		"hl_rsa3072"
-
-enum fit_load_result {
-	FIT_LOAD_SUCCESS = 0,
-	FIT_LOAD_CORRUPTED,
-	FIT_LOAD_BAD_LOAD_ADDR,
-	FIT_LOAD_NOT_RUNNABLE,
-	FIT_LOAD_SEC_FAILED,
-};
-
-enum fit_format_result {
-	FIT_FORMAT_SUCCESS = 0,
-	FIT_FORMAT_HEADER_CHECK_FAILED,
-	FIT_FORMAT_NO_DESCRIPTION,
-	FIT_FORMAT_NO_IMAGES_PARENT_NODE,
-	FIT_FORMAT_CURRUPTED_LOAD_TYPE,
-	FIT_FORMAT_UNSUPPORTED_LOAD_TYPE,
-};
-
-void fit_load_print_error(enum fit_load_result errtype);
-void fit_format_print_error(enum fit_format_result errtype);
-int fit_load_images(void *fit);
-void fit_run_image(void);
-int fit_validate_size(const void *fit, int len);
-int fit_get_load_type(const void *fit, u32 *load_type);
-int fit_image_get_loadable_data(const void *fit, const void **data,
-				size_t size);
-int fit_init_resp(void *fit, size_t sz);
-int fit_fin_resp(void *fit, unsigned int *res_size);
-int fit_verify_img_sig(const void *img, size_t img_size, const void *sec_desc,
-		       size_t sec_desc_size, const void *pub_key_fit,
-		       const u8 *sig, size_t sig_len);
-int fit_process_core_id(enum fit_core_id core_id);
-bool fit_is_core_id_supported(void);
-
-#endif /* _HL_FIT_H_ */
diff --git a/lib/preboot/fit_regions.c b/lib/preboot/fit_regions.c
deleted file mode 100644
index e89e9e6..0000000
--- a/lib/preboot/fit_regions.c
+++ /dev/null
@@ -1,168 +0,0 @@
-// SPDX-License-Identifier:     GPL-2.0+
-/*
- * Copyright (C) 2020 HabanaLabs Ltd.
- */
-
-#include <libfdt.h>
-#include "common.h"
-#include "fit_rsa.h"
-#include "fit_regions.h"
-#include "sha.h"
-
-#define FDT_MAX_DEPTH	8
-
-/*
- * from u-boot/common/image-sig.c
- */
-struct image_region *fit_region_make_list(const void *fit,
-					  struct fdt_region *fdt_regions,
-					  int count,
-					  struct image_region *region)
-{
-	int i;
-
-	for (i = 0; i < count; i++) {
-		region[i].data = fit + fdt_regions[i].offset;
-		region[i].size = fdt_regions[i].size;
-	}
-
-	return region;
-}
-
-static int str_in_list(const char *str, char * const list[], int count)
-{
-	int i;
-
-	for (i = 0; i < count; i++)
-		if (!strcmp(list[i], str))
-			return 1;
-
-	return 0;
-}
-
-/*
- * from u-boot/lib/libfdt/fdt_region.c
- */
-int fdt_find_regions(const void *fdt, char * const inc[], int inc_count,
-		     char * const exc_prop[], int exc_prop_count,
-		     struct fdt_region region[], int max_regions,
-		     char *path, int path_len, int add_string_tab)
-{
-	int stack[FDT_MAX_DEPTH] = { 0 };
-	char *end;
-	int nextoffset = 0;
-	u32 tag;
-	int count = 0;
-	int start = -1;
-	int depth = -1;
-	int want = 0;
-	int base = fdt_off_dt_struct(fdt);
-
-	end = path;
-	*end = '\0';
-	do {
-		const struct fdt_property *prop;
-		const char *name;
-		const char *str;
-		int include = 0;
-		int stop_at = 0;
-		int offset;
-		int len;
-
-		offset = nextoffset;
-		tag = fdt_next_tag(fdt, offset, &nextoffset);
-		stop_at = nextoffset;
-
-		switch (tag) {
-		case FDT_PROP:
-			include = want >= 2;
-			stop_at = offset;
-			prop = fdt_get_property_by_offset(fdt, offset, NULL);
-			str = fdt_string(fdt, fdt32_to_cpu(prop->nameoff));
-			if (str_in_list(str, exc_prop, exc_prop_count))
-				include = 0;
-			break;
-
-		case FDT_NOP:
-			include = want >= 2;
-			stop_at = offset;
-			break;
-
-		case FDT_BEGIN_NODE:
-			depth++;
-			if (depth == FDT_MAX_DEPTH)
-				return -FDT_ERR_BADSTRUCTURE;
-			name = fdt_get_name(fdt, offset, &len);
-			if (end - path + 2 + len >= path_len)
-				return -FDT_ERR_NOSPACE;
-			if (end != path + 1)
-				*end++ = '/';
-			memcpy(end, name, len+1);
-			end += len;
-			stack[depth] = want;
-			if (want == 1)
-				stop_at = offset;
-			if (str_in_list(path, inc, inc_count))
-				want = 2;
-			else if (want)
-				want--;
-			else
-				stop_at = offset;
-			include = want;
-			break;
-
-		case FDT_END_NODE:
-			/* Depth must never go below -1 */
-			if (depth < 0)
-				return -FDT_ERR_BADSTRUCTURE;
-			include = want;
-			want = stack[depth--];
-			while (end > path && *--end != '/')
-				;
-			*end = '\0';
-			break;
-
-		case FDT_END:
-			include = 1;
-			break;
-		}
-
-		if (include && start == -1) {
-			/* Should we merge with previous? */
-			if (count && count <= max_regions &&
-			    offset == region[count - 1].offset +
-					region[count - 1].size - base)
-				start = region[--count].offset - base;
-			else
-				start = offset;
-		}
-
-		if (!include && start != -1) {
-			if (count < max_regions) {
-				region[count].offset = base + start;
-				region[count].size = stop_at - start;
-			}
-			count++;
-			start = -1;
-		}
-	} while (tag != FDT_END);
-
-	if (nextoffset != fdt_size_dt_struct(fdt))
-		return -FDT_ERR_BADLAYOUT;
-
-	/* Add a region for the END tag and the string table */
-	if (count < max_regions) {
-		region[count].offset = base + start;
-		region[count].size = nextoffset - start;
-		if (add_string_tab)
-			region[count].size += fdt_size_dt_strings(fdt);
-	}
-	count++;
-
-	return count;
-}
-
-__weak void sha_csum_regions(const struct image_region region[],
-				int region_count, u8 *output)
-{
-}
diff --git a/lib/preboot/fit_regions.h b/lib/preboot/fit_regions.h
deleted file mode 100644
index 0254526..0000000
--- a/lib/preboot/fit_regions.h
+++ /dev/null
@@ -1,24 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2020 HabanaLabs Ltd.
- */
-
-#ifndef COMMON_FIT_REGIONS_H_
-#define COMMON_FIT_REGIONS_H_
-
-struct fdt_region {
-	int offset;
-	int size;
-};
-
-struct image_region *fit_region_make_list(const void *fit,
-					  struct fdt_region *fdt_regions,
-					  int count,
-					  struct image_region *region);
-
-int fdt_find_regions(const void *fdt, char * const inc[], int inc_count,
-		     char * const exc_prop[], int exc_prop_count,
-		     struct fdt_region region[], int max_regions,
-		     char *path, int path_len, int add_string_tab);
-
-#endif /* COMMON_FIT_REGIONS_H_ */
diff --git a/lib/preboot/fit_rsa.c b/lib/preboot/fit_rsa.c
deleted file mode 100644
index 036bb27..0000000
--- a/lib/preboot/fit_rsa.c
+++ /dev/null
@@ -1,360 +0,0 @@
-// SPDX-License-Identifier:     GPL-2.0+
-/*
- * Copyright (C) 2017 HabanaLabs Ltd.
- */
-
-#include <libfdt.h>
-#include "common.h"
-#include "uart.h"
-#include "fit_rsa.h"
-#include "fit_regions.h"
-#include "sha.h"
-
-struct rsa_public_key {
-	u32 len;	/* len of modulus[] in number of u32 */
-	u32 n0inv;	/* -1 / modulus[0] mod 2^32 */
-	u32 *modulus;	/* modulus as little endian array */
-	u32 *rr;	/* R^2 as little endian array */
-	u64 exponent;	/* public exponent */
-};
-
-#define CHECKSUM_LENGTH		(SHA_SUM_LEN)
-#define CHECKSUM_DER_LENGTH	(SHA_DER_LEN)
-#define CHECKSUM_DER_PREFIX	(SHA_DER_PRE)
-
-
-#define RSA4096_BITS		(4096)
-#define RSA4096_BYTES		(RSA4096_BITS / 8)
-#define RSA3072_BITS		(3072)
-#define RSA3072_BYTES		(RSA3072_BITS / 8)
-#define CRYPTO_RSA_LENGTH	(RSA3072_BYTES)
-#define RSA_DEFAULT_PUBEXP	(65537)
-
-#if (CHECKSUM_LENGTH > CRYPTO_RSA_LENGTH)
-#error "CHECKSUM_LENGTH > CRYPTO_RSA_LENGTH"
-#endif
-
-#define get_unaligned_be32(a) fdt32_to_cpu(*(u32 *)a)
-#define put_unaligned_be32(a, b) (*(u32 *)(b) = cpu_to_fdt32(a))
-
-
-/**
- * is_public_exponent_bit_set() - Check if a bit in the public exponent is set
- *
- * @key:	RSA key
- * @pos:	The bit position to check
- */
-static int is_public_exponent_bit_set(const struct rsa_public_key *key,
-				      int pos)
-{
-	return key->exponent & (1ULL << pos);
-}
-
-/**
- * subtract_modulus() - subtract modulus from the given value
- *
- * @key:	Key containing modulus to subtract
- * @num:	Number to subtract modulus from, as little endian word array
- */
-static void subtract_modulus(const struct rsa_public_key *key, u32 num[])
-{
-	long long acc = 0;
-	u32 i;
-
-	for (i = 0; i < key->len; i++) {
-		acc += (u64)num[i] - key->modulus[i];
-		num[i] = (u32)acc;
-		acc >>= 32;
-	}
-}
-
-
-/**
- * greater_equal_modulus() - check if a value is >= modulus
- *
- * @key:	Key containing modulus to check
- * @num:	Number to check against modulus, as little endian word array
- * @return 0 if num < modulus, 1 if num >= modulus
- */
-static int greater_equal_modulus(const struct rsa_public_key *key,
-				 u32 num[])
-{
-	int i;
-
-	for (i = (int)key->len - 1; i >= 0; i--) {
-		if (num[i] < key->modulus[i])
-			return 0;
-		if (num[i] > key->modulus[i])
-			return 1;
-	}
-
-	return 1;	/* equal */
-}
-
-/**
- * num_pub_exponent_bits() - Number of bits in the public exponent
- *
- * @key:	RSA key
- * @num_bits:	Storage for the number of public exponent bits
- */
-static int num_public_exponent_bits(const struct rsa_public_key *key,
-				    int *num_bits)
-{
-	u64 exponent;
-	int exponent_bits;
-	const u32 max_bits = (sizeof(exponent) * 8);
-
-	exponent = key->exponent;
-	exponent_bits = 0;
-
-	if (!exponent) {
-		*num_bits = exponent_bits;
-		return 0;
-	}
-
-	for (exponent_bits = 1; exponent_bits < max_bits + 1; ++exponent_bits)
-		if (!(exponent >>= 1)) {
-			*num_bits = exponent_bits;
-			return 0;
-		}
-
-	return 1;
-}
-
-/**
- * montgomery_mul_add_step() - Perform montgomery multiply-add step
- *
- * Operation: montgomery result[] += a * b[] / n0inv % modulus
- *
- * @key:	RSA key
- * @result:	Place to put result, as little endian word array
- * @a:		Multiplier
- * @b:		Multiplicand, as little endian word array
- */
-static void montgomery_mul_add_step(const struct rsa_public_key *key,
-				    u32 result[], const u32 a, const u32 b[])
-{
-	u64 acc_a, acc_b;
-	u32 d0;
-	u32 i;
-
-	acc_a = (u64)a * b[0] + result[0];
-	d0 = (u32)acc_a * key->n0inv;
-	acc_b = (u64)d0 * key->modulus[0] + (u32)acc_a;
-	for (i = 1; i < key->len; i++) {
-		acc_a = (acc_a >> 32) + (u64)a * b[i] + result[i];
-		acc_b = (acc_b >> 32) + (u64)d0 * key->modulus[i] +
-				(u32)acc_a;
-		result[i - 1] = (u32)acc_b;
-	}
-
-	acc_a = (acc_a >> 32) + (acc_b >> 32);
-
-	result[i - 1] = (u32)acc_a;
-
-	if (acc_a >> 32)
-		subtract_modulus(key, result);
-}
-
-/**
- * montgomery_mul() - Perform montgomery mutitply
- *
- * Operation: montgomery result[] = a[] * b[] / n0inv % modulus
- *
- * @key:	RSA key
- * @result:	Place to put result, as little endian word array
- * @a:		Multiplier, as little endian word array
- * @b:		Multiplicand, as little endian word array
- */
-static void montgomery_mul(const struct rsa_public_key *key,
-			   u32 result[], u32 a[], const u32 b[])
-{
-	u32 i;
-
-	for (i = 0; i < key->len; ++i)
-		result[i] = 0;
-	for (i = 0; i < key->len; ++i)
-		montgomery_mul_add_step(key, result, a[i], b);
-}
-
-static int pow_mod(const struct rsa_public_key *key, u32 *out, u32 *in)
-{
-	u32 *result, *ptr;
-	u32 i;
-	int j, k;
-
-	u32 val[key->len], acc[key->len], tmp[key->len];
-	u32 a_scaled[key->len];
-
-	result = tmp;	/* Re-use location. */
-
-	/* Convert from big endian byte array to little endian word array. */
-	for (i = 0, ptr = in + key->len - 1; i < key->len; i++, ptr--)
-		val[i] = get_unaligned_be32(ptr);
-
-	if (num_public_exponent_bits(key, &k) != 0) {
-		preboot_puts("num_public_exponent_bits failed!\n");
-		return 1;
-	}
-
-	if (k < 2) {
-		preboot_puts("Public exponent is too short\n");
-		return 1;
-	}
-
-	if (!is_public_exponent_bit_set(key, 0)) {
-		preboot_puts("LSB of RSA public exponent must be set.\n");
-		return 1;
-	}
-
-	/* the bit at e[k-1] is 1 by definition, so start with: C := M */
-	montgomery_mul(key, acc, val, key->rr); /* acc = a * RR / R mod n */
-	/* retain scaled version for intermediate use */
-	memcpy(a_scaled, acc, key->len * sizeof(a_scaled[0]));
-
-	for (j = k - 2; j > 0; --j) {
-		montgomery_mul(key, tmp, acc, acc); /* tmp = acc^2 / R mod n */
-
-		if (is_public_exponent_bit_set(key, j)) {
-			/* acc = tmp * val / R mod n */
-			montgomery_mul(key, acc, tmp, a_scaled);
-		} else {
-			/* e[j] == 0, copy tmp back to acc for next operation */
-			memcpy(acc, tmp, key->len * sizeof(acc[0]));
-		}
-	}
-
-	/* the bit at e[0] is always 1 */
-	montgomery_mul(key, tmp, acc, acc); /* tmp = acc^2 / R mod n */
-	montgomery_mul(key, acc, tmp, val); /* acc = tmp * a / R mod M */
-	memcpy(result, acc, key->len * sizeof(result[0]));
-
-	/* Make sure result < mod; result is at most 1x mod too large. */
-	if (greater_equal_modulus(key, result))
-		subtract_modulus(key, result);
-
-	/* Convert to bigendian byte array */
-	for (i = key->len - 1, ptr = out; (int)i >= 0; i--, ptr++)
-		put_unaligned_be32(result[i], ptr);
-	return 0;
-}
-
-static void rsa_convert_big_endian(u32 *dst, const u32 *src, int len)
-{
-	int i;
-
-	for (i = 0; i < len; i++)
-		dst[i] = fdt32_to_cpu(src[len - 1 - i]);
-}
-
-static int rsa_mod_exp_sw(const u8 *sig, struct rsa_key_prop *prop, u32 *out)
-{
-	struct rsa_public_key key;
-	int ret;
-
-	key.n0inv = prop->n0inv;
-	key.len = prop->num_bits / (sizeof(u32) * 8);
-	if (prop->public_exponent) {
-		u32 *pubexp = (u32 *)&(key.exponent);
-
-		/* the value property is 4 bytes align. trying to read 64bit
-		 * value aligned to 4bytes by fdt64_to_cpu lead to crash
-		 */
-		pubexp[0] = fdt32_to_cpu(((u32 *)prop->public_exponent)[1]);
-		pubexp[1] = fdt32_to_cpu(((u32 *)prop->public_exponent)[0]);
-	} else {
-		key.exponent = RSA_DEFAULT_PUBEXP;
-	}
-
-	u32 key1[key.len], key2[key.len];
-
-	key.modulus = key1;
-	key.rr = key2;
-
-	rsa_convert_big_endian(key.modulus, (u32 *)prop->modulus, key.len);
-	rsa_convert_big_endian(key.rr, (u32 *)prop->rr, key.len);
-
-	ret = pow_mod(&key, out, (u32 *)sig);
-	if (ret)
-		return ret;
-
-	return 0;
-}
-
-static int rsa_verify_padding(const u8 *msg, const int pad_len)
-{
-	int ff_len;
-	int ret;
-
-	/* first byte must be 0x00 */
-	ret = *msg++;
-	/* second byte must be 0x01 */
-	ret |= *msg++ ^ 0x01;
-	/* next ff_len bytes must be 0xff */
-	ff_len = pad_len - CHECKSUM_DER_LENGTH - 3;
-	ret |= *msg ^ 0xff;
-	ret |= memcmp(msg, msg+1, ff_len-1);
-	msg += ff_len;
-	/* next byte must be 0x00 */
-	ret |= *msg++;
-	/* next der_len bytes must match der_prefix */
-	ret |= memcmp(msg, CHECKSUM_DER_PREFIX, CHECKSUM_DER_LENGTH);
-
-	return ret;
-}
-
-int padding_pkcs_15_verify(u8 *msg, int msg_len, const u8 *hash)
-{
-	int pad_len = msg_len - CHECKSUM_LENGTH;
-
-	if (rsa_verify_padding(msg, pad_len)) {
-		preboot_puts("RSAVerify(): rsa_verify_padding failed!\n");
-		return 1;
-	}
-
-	/* Check hash. */
-	if (memcmp((u8 *)msg + pad_len, hash, msg_len - pad_len)) {
-		preboot_puts("RSAVerify(): Hash check failed!\n");
-		return 1;
-	}
-
-	return 0;
-}
-
-static int rsa_verify_key(struct rsa_key_prop *prop, const u8 *sig,
-			  const u32 sig_len, const u8 *hash, const u32 key_len)
-{
-	u32 buf[sig_len>>2];
-
-	if (rsa_mod_exp_sw(sig, prop, buf)) {
-		preboot_puts("rsa_mod_exp_sw failed\n");
-		return 1;
-	}
-
-	if (padding_pkcs_15_verify((u8 *)buf, key_len, hash))
-		return 1;
-
-	return 0;
-}
-
-/* based on int rsa_verify(struct image_sign_info *info,
- *		const struct image_region region[], int region_count,
- *		uint8_t *sig, uint sig_len)
- * from: u-boot/lib/rsa/rsa-verify.c
- */
-int rsa_verify_regions(const struct image_region region[], int region_count,
-		       const u8 *sig, u32 sig_len, struct rsa_key_prop *prop)
-{
-	u32 hash[CRYPTO_RSA_LENGTH >> 2];
-	int ret;
-
-	if (sig_len != (prop->num_bits >> 3)) {
-		preboot_puts("sig_len != (prop->num_bits / 8)\n");
-		return 1;
-	}
-
-	sha_csum_regions(region, region_count, (u8 *)hash);
-	ret = rsa_verify_key(prop, sig, sig_len, (u8 *)hash, CRYPTO_RSA_LENGTH);
-	preboot_puts_hex("rsa_verify_regions ret: ", ret, 0);
-	return ret;
-}
diff --git a/lib/preboot/fit_rsa.h b/lib/preboot/fit_rsa.h
deleted file mode 100644
index be24bc1..0000000
--- a/lib/preboot/fit_rsa.h
+++ /dev/null
@@ -1,30 +0,0 @@
-/* SPDX-License-Identifier:     GPL-2.0+
- *
- * Copyright (C) 2017 HabanaLabs Ltd.
- */
-
-#ifndef COMMON_FIT_RSA_H_
-#define COMMON_FIT_RSA_H_
-
-#include <baremetal.h>
-
-struct rsa_key_prop {
-	const void *rr;			/* R^2 can be treated as byte array */
-	const void *modulus;		/* modulus as byte array */
-	const u64 *public_exponent;	/* public exponent as byte array */
-	u32 n0inv;			/* -1 / modulus[0] mod 2^32 */
-	int num_bits;			/* Key length in bits */
-};
-
-struct image_region {
-	const void *data;
-	int size;
-};
-
-
-int rsa_verify_regions(const struct image_region region[], int region_count,
-		       const u8 *sig, u32 sig_len, struct rsa_key_prop *prop);
-int fit_get_key_prop(const void *pub_key_fit, const char *key_name,
-			    struct rsa_key_prop *prop);
-
-#endif /* COMMON_FIT_RSA_H_ */
diff --git a/lib/preboot/i2cslave_indirect.c b/lib/preboot/i2cslave_indirect.c
index f7d6319..04caacf 100644
--- a/lib/preboot/i2cslave_indirect.c
+++ b/lib/preboot/i2cslave_indirect.c
@@ -185,6 +185,7 @@
 static void i2csi_req_load_fw(void)
 {
 	struct mem_desc desc;
+	struct fit_error fit_err = {};
 	/*
 	 * Received firmware length is the last requset RDQL + previous
 	 * requests length which is calculate by the header offset.
@@ -198,7 +199,7 @@
 	if (fit_validate_size((void *)desc.base, fw_len))
 		return; /* More data is needed */
 
-	if (fit_load_images((void *)desc.base))
+	if (fit_load_images(((void *)desc.base), &fit_err))
 		req_sts.err = HL_I2CI_REQ_ERR_PROCESS;
 	else
 		i2cslave_fit_set_checked();
diff --git a/lib/preboot/module_fwupd.c b/lib/preboot/module_fwupd.c
index fa78557..edb4e58 100644
--- a/lib/preboot/module_fwupd.c
+++ b/lib/preboot/module_fwupd.c
@@ -24,6 +24,7 @@
 #include "uart.h"
 #include "pll.h"
 #include "fit.h"
+#include <fit_load.h>
 
 #define WFC_TIMEOUT_MS			500
 #define TIMEOUT_US			1000
@@ -639,32 +640,9 @@
 
 __weak void preboot_nic_en(u32 nic_en) {}
 
-static int module_fwupd_load_type_nic_en(void *fit)
+static void module_fwupd_load_type_run(struct fwupd_state *state)
 {
-	int err;
-	const void *data = 0;
-
-	err = fit_image_get_loadable_data(fit, &data, sizeof(u32));
-	if (!err)
-		preboot_nic_en(*(u32 *)data);
-
-	return err;
-}
-
-__weak int rma_erase(void)
-{
-	return 1;
-}
-
-static int module_fwupd_load_type_run(struct fwupd_state *state, void *fit)
-{
-	int err;
-
 	state->exec_res = NOT_FOUND;
-
-	err = fit_load_images(fit);
-	if (err)
-		return err;
 
 	if (state->synced_no_return) {
 		state->sync_exec_res = FOUND_NO_RETURN;
@@ -672,37 +650,6 @@
 	} else {
 		state->exec_res = FOUND_NO_RETURN;
 	}
-
-	return err;
-}
-
-static int module_fwupd_load_type(struct fwupd_state *state)
-{
-	u32 load_type;
-	int ret;
-
-	ret = fit_get_load_type(state->map_offset, &load_type);
-	if (ret) {
-		fit_format_print_error(ret);
-		return 1;
-	}
-
-	switch (load_type) {
-	case FIT_LOAD_TYPE_NIC_EN_MASK:
-		return module_fwupd_load_type_nic_en(state->map_offset);
-	case FIT_LOAD_TYPE_RMA_ERASE:
-		return rma_erase();
-	case FIT_LOAD_TYPE_FW_UPD_CLOSURE:
-		module_fwupd_pcie_disable(true);
-		break;
-	case FIT_LOAD_TYPE_RUN:
-		return module_fwupd_load_type_run(state, state->map_offset);
-	default:
-		preboot_puts("Unknown FIT load type\n");
-		return 1;
-	}
-
-	return 0;
 }
 
 static enum hl_upd_sts module_fwupd_run_cmd(struct fwupd_state *state,
@@ -711,6 +658,8 @@
 {
 	int err;
 	enum hl_upd_sts sts = HL_UPD_STS_OK;
+	bool is_secured, is_run = false;
+	struct fit_error fit_err = {};
 
 	/*
 	 * we do not allow mapping of SRAM and SPSRAM at the same time -
@@ -720,7 +669,14 @@
 	preboot_security_dynamic_img_sram_close();
 	preboot_security_dynamic_img_elbi_reg_close();
 
-	err = module_fwupd_load_type(state);
+	is_secured = state->use_security;
+	err = fit_verify(state->map_offset, &fit_err);
+	if (!err)
+		err = fit_load_type(state->map_offset, is_secured, &is_run, &fit_err);
+
+	fit_report_error(&fit_err);
+	if (is_run)
+		module_fwupd_load_type_run(state);
 
 	if (cmd == HL_UPD_RUN_ELBI)
 		preboot_dynamic_img_elbi_reg_cleanup(state->map_offset, ROUND_UP4(state->map_size));
